<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>&lt;no title&gt; &#8212; cogent3 2020.4 documentation</title>
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Project History" href="history.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="#">
          cogent3</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="install.html">Install</a></li>
                <li><a href="https://cogent3.readthedocs.io">Docs</a></li>
                <li><a href="https://github.com/cogent3/cogent3">Code</a></li>
                <li><a href="projects.html">Projects Using</a></li>
                <li><a href="https://github.com/cogent3/c3dev/wiki/Code-of-Conduct">Code of Conduct</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference external" href="https://github.com/cogent3/">Documentation Release</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/cogent3/cetest">Documentation Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">Project History</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"></ul>
</li>
              
            
            
              
                
  <li>
    <a href="history.html" title="Next Chapter: Project History"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">Project History &raquo;</span>
    </a>
  </li>
              
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <p><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> is a mature python library for analysis of biological sequence data. We endeavour to provide a first-class experience within Jupyter notebooks, but the algorithms also support parallel execution on compute systems with 1000’s of processors.</p>
<p><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> is released under the BSD-3 license, links to documentation and code are above. If you would like to contribute (and we hope you do!), we have created a companion <a class="reference external" href="https://github.com/cogent3/c3dev">c3dev</a> repo which provides details on how to contribute and some useful tools for doing so.</p>
<p><strong>Who is it for? Anyone who wants to analyse sequence divergence using robust statistical models.</strong></p>
<p><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> is unique in providing numerous non-stationary Markov models <a class="bibtex reference internal" href="#kaehler-000" id="id1">[1]</a> for modelling sequence evolution, including novel codon models <a class="bibtex reference internal" href="#id11" id="id2">[2]</a>. <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> also includes an extensive collection of time-reversible models (again including novel codon models  <a class="bibtex reference internal" href="#yap-000" id="id3">[3]</a>). We have done more than just invent these new methods, we have established the most robust algorithms <a class="bibtex reference internal" href="#schranz-000" id="id4">[4]</a> for their implementation and their suitability for real data <a class="bibtex reference internal" href="#verbyla-000" id="id5">[5]</a>. Additionally, there are novel signal processing methods focussed on statistical estimation of integer period signals <a class="bibtex reference internal" href="#epps-000" id="id6">[6]</a><a class="bibtex reference internal" href="#bellani-000" id="id7">[7]</a>.</p>
<div class="admonition-todo admonition" id="id8">
<p class="admonition-title">Todo</p>
<p>LINK TO DOCS</p>
</div>
<p><strong>Anyone who wants to undertake exploratory genomic data analysis</strong></p>
<p>Beyond our novel methods, <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> provides an extensive suite of capabilities for manipulating and analysing sequence data. For instance, the ability to read standard biological data formats, manipulate sequences by their annotations, to perform multiple sequence alignment using any of our substitution models, phylogenetic reconstruction and tree manipulation, manipulation of tabular data, visualisation of phylogenies and much more.</p>
<div class="admonition-todo admonition" id="id9">
<p class="admonition-title">Todo</p>
<p>LINK TO DOCS</p>
</div>
<p><strong>Anyone looking for a functional programming style approach to genomic data analysis</strong></p>
<p>Our <code class="docutils literal notranslate"><span class="pre">cogent3.app</span></code> module provides a very different approach to using the library capabilities. Notably, a functional programming style interface lowers the barrier to entry for using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code>’s advanced capabilities. It also supports building pipelines suitable for large-scale analysis. Individuals comfortable with R should find this interface pretty easy to use.</p>
<div class="admonition-todo admonition" id="id10">
<p class="admonition-title">Todo</p>
<p>LINK TO DOCS</p>
</div>
<div class="toctree-wrapper compound">
</div>
<hr class="docutils" />
<p class="rubric">References</p>
<p id="bibtex-bibliography-index-0"><dl class="citation">
<dt class="bibtex label" id="kaehler-000"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Benjamin D Kaehler, Von Bing Yap, Rongli Zhang, and Gavin A Huttley. Genetic distance for a general non-stationary Markov substitution process. <em>Systematic Biology</em>, 64(2):281–93, mar 2015. URL: <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/25503772">http://www.ncbi.nlm.nih.gov/pubmed/25503772</a>.</p>
</dd>
<dt class="bibtex label" id="id11"><span class="brackets"><a class="fn-backref" href="#id2">2</a></span></dt>
<dd><p>Benjamin D Kaehler, Von Bing Yap, and Gavin A Huttley. Standard Codon Substitution Models Overestimate Purifying Selection for Non-Stationary Data. <em>Genome Biology and Evolution</em>, 9(1):134–149, jan 2017. URL: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/28175284">https://www.ncbi.nlm.nih.gov/pubmed/28175284</a>.</p>
</dd>
<dt class="bibtex label" id="yap-000"><span class="brackets"><a class="fn-backref" href="#id3">3</a></span></dt>
<dd><p>Von Bing Yap, Helen Lindsay, Simon Easteal, and Gavin Huttley. Estimates of the Effect of Natural Selection on Protein-Coding Content. <em>Molecular Biology and Evolution</em>, 27(3):726–734, mar 2010. URL: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/19815689">https://www.ncbi.nlm.nih.gov/pubmed/19815689</a>.</p>
</dd>
<dt class="bibtex label" id="schranz-000"><span class="brackets"><a class="fn-backref" href="#id4">4</a></span></dt>
<dd><p>Harold W Schranz, Von Bing Yap, Simon Easteal, Rob Knight, and Gavin A Huttley. Pathological rate matrices: from primates to pathogens. <em>BMC Bioinformatics</em>, 9:550, 2008. URL: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/19099591">https://www.ncbi.nlm.nih.gov/pubmed/19099591</a>.</p>
</dd>
<dt class="bibtex label" id="verbyla-000"><span class="brackets"><a class="fn-backref" href="#id5">5</a></span></dt>
<dd><p>Klara L Verbyla, Von Bing Yap, Anuj Pahwa, Yunli Shao, and Gavin A Huttley. The Embedding Problem for Markov Models of Nucleotide Substitution. <em>PLoS ONE</em>, 8(7):e69187, 2013. URL: <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/23935949/">https://pubmed.ncbi.nlm.nih.gov/23935949/</a>.</p>
</dd>
<dt class="bibtex label" id="epps-000"><span class="brackets"><a class="fn-backref" href="#id6">6</a></span></dt>
<dd><p>Julien Epps, Hua Ying, and Gavin Huttley. Statistical methods for detecting periodic fragments in DNA sequence data. <em>Biology Direct</em>, 6(1):21, 2011. URL: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/21527008">https://www.ncbi.nlm.nih.gov/pubmed/21527008</a>.</p>
</dd>
<dt class="bibtex label" id="bellani-000"><span class="brackets"><a class="fn-backref" href="#id7">7</a></span></dt>
<dd><p>M Bellani, J Epps, and G Huttley. A Comparison of Periodicity Profile Methods for Sequence Analysis. <em>Proc. IEEE Workshop on Genomic Signal Processing and Statistics</em>, 2012. URL: <a class="reference external" href="https://ieeexplore.ieee.org/document/6507731">https://ieeexplore.ieee.org/document/6507731</a>.</p>
</dd>
</dl>
</p>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020, Gavin Huttley.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.0.2.<br/>
    </p>
  </div>
</footer>
  </body>
</html>