<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>What apps are there? &#8212; cogent3 2020.4 documentation</title>
    <link rel="stylesheet" href="../../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-dropdown.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/language_data.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../_static/require.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../../_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">
          cogent3</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../install.html">Install</a></li>
                <li><a href="../index.html">Docs</a></li>
                <li><a href="../draw/index.html">Gallery</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Project History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pycogent.html">How does <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> relate to PyCogent?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">What apps are there?</a></li>
</ul>
</ul>
</li>
              
            
            
              
                
              
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="section" id="what-apps-are-there">
<h1>What apps are there?<a class="headerlink" href="#what-apps-are-there" title="Permalink to this headline">¶</a></h1>
<p>Apps perform functions ranging from multiple sequence alignment (e.g. <code class="docutils literal notranslate"><span class="pre">progressive_align</span></code>), to excluding alignment columns containing non-nucleotide characters (e.g. <code class="docutils literal notranslate"><span class="pre">omit_degenerates</span></code>) to performing maximum-likelihood evolutionary analyses (e.g. <code class="docutils literal notranslate"><span class="pre">model</span></code>).</p>
<p>Apps that are identified as “composable” (with a value of <code class="docutils literal notranslate"><span class="pre">True</span></code> under that column) can be combined into a single function by addition.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">available_apps</span>

<span class="n">available_apps</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>module</th>
<th>name</th>
<th>composable</th>
<th>doc</th>
<th>inputs</th>
<th>outputs</th>
<th>data type</th>
</thead>
<tbody>
<tr>
<td>cogent3.app.align</td>
<td>align_to_ref</td>
<td>True</td>
<td>Aligns to a reference seq, no gaps in the reference.
    Returns an Alignment object.</td>
<td>sequences</td>
<td>aligned, serialisable</td>
<td>SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.align</td>
<td>progressive_align</td>
<td>True</td>
<td>Progressive multiple sequence alignment via any cogent3 model.
     Returns an Alignment object.</td>
<td>sequences</td>
<td>aligned, serialisable</td>
<td>SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.dist</td>
<td>fast_slow_dist</td>
<td>True</td>
<td>Pairwise distance calculation. Uses fast (but less
    numerically robust) approach where possible, slow (robust)
    approach when not. Returns a DistanceMatrix.</td>
<td>aligned</td>
<td>pairwise_distances, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>ancestral_states</td>
<td>True</td>
<td>Computes ancestral state probabilities from a model result. Returns a dict
    with a DictArray for each node.</td>
<td>model_result</td>
<td>result, tabular_result, serialisable</td>
<td>model_result</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>bootstrap</td>
<td>True</td>
<td>Parametric bootstrap for a provided hypothesis. Returns a bootstrap_result.</td>
<td>aligned</td>
<td>result, bootstrap_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>hypothesis</td>
<td>True</td>
<td>Specify a hypothesis through defining two models. Returns a
    hypothesis_result.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>model</td>
<td>True</td>
<td>Define a substitution model + tree for maximum likelihood evaluation.
    Returns model_result.</td>
<td>aligned, serialisable</td>
<td>result, model_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>natsel_neutral</td>
<td>True</td>
<td>Test of selective neutrality by assessing whether omega equals 1.
    Under the alternate, there is one omega for all branches and all sites.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>natsel_sitehet</td>
<td>True</td>
<td>Test for site-heterogeneity in omega. Under null, there are 2 site-classes,
    omega &lt; 1 and omega = 1. Under the alternate, an additional site-class of
    omega &gt; 1 is added.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>natsel_timehet</td>
<td>True</td>
<td>The branch heterogeneity hypothesis test for natural selection.
    Tests for whether a single omega for all branches is sufficient against the
    alternate that a user specified subset of branches have a distinct value
    (or values) of omega.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>natsel_zhang</td>
<td>True</td>
<td>The branch by site-class hypothesis test for natural selection of
    Zhang et al MBE 22: 2472-2479.

    Note: Our implementation is not as parametrically succinct as that of
    Zhang et al, we have 1 additional bin probability.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>tabulate_stats</td>
<td>True</td>
<td>Extracts all model statistics from model_result as Table.</td>
<td>model_result, serialisable</td>
<td>result, tabular_result, serialisable</td>
<td>model_result</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_aligned</td>
<td>True</td>
<td>Loads aligned sequences. Returns an Alignment object.</td>
<td></td>
<td>aligned, serialisable</td>
<td>DataStoreMember, str, Path</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_db</td>
<td>True</td>
<td>Loads json serialised cogent3 objects from a TinyDB file. 
    Returns whatever object type was stored.</td>
<td></td>
<td>serialisable</td>
<td></td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_json</td>
<td>True</td>
<td>Loads json serialised cogent3 objects from a json file. 
    Returns whatever object type was stored.</td>
<td></td>
<td>serialisable</td>
<td></td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_tabular</td>
<td>True</td>
<td>Loads delimited data. Returns a Table.</td>
<td></td>
<td>tabular, serialisable</td>
<td>DataStoreMember, str, Path</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_unaligned</td>
<td>True</td>
<td>Loads unaligned sequences. Returns a SequenceCollection.</td>
<td></td>
<td>sequences, serialisable</td>
<td>DataStoreMember, str, Path</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>write_db</td>
<td>True</td>
<td>Writes json serialised objects to a TinyDB instance.</td>
<td>serialisable</td>
<td>identifier, serialisable</td>
<td></td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>write_json</td>
<td>True</td>
<td>Writes json serialised objects to individual json files.</td>
<td>serialisable</td>
<td>identifier, serialisable</td>
<td></td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>write_seqs</td>
<td>True</td>
<td>Writes sequences to text files in standard format.</td>
<td>sequences, aligned</td>
<td>sequences, aligned, identifier</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>write_tabular</td>
<td>True</td>
<td>writes tabular data</td>
<td>tabular_result, tabular</td>
<td>identifier</td>
<td>Table, DictArray, DistanceMatrix</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>concat</td>
<td>False</td>
<td>Creates a concatenated alignment from a series. Returns an Alignment.</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>fixed_length</td>
<td>True</td>
<td>Sample an alignment to a fixed length. Returns an Alignment of the 
    specified length, or NotCompleted if alignment too short.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>min_length</td>
<td>True</td>
<td>Filters sequence collections / alignments by length. Returns the 
    data if it satisfies the condition, NotCompleted otherwise.</td>
<td>sequences, aligned, serialisable</td>
<td>sequences, aligned, serialisable</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>omit_bad_seqs</td>
<td>True</td>
<td>Eliminates sequences from Alignment based on gap fraction, unique gaps.
    Returns modified alignment.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>omit_degenerates</td>
<td>True</td>
<td>Excludes alignment columns with degenerate conditions. Can accomodate
    reading frame. Returns modified Alignment.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>omit_duplicated</td>
<td>True</td>
<td>Removes redundant sequences, recording dropped sequences in
    seqs.info.dropped. Returns sequence collection with only unique sequences.</td>
<td>sequences, serialisable, aligned</td>
<td>sequences, serialisable, aligned</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>omit_gap_pos</td>
<td>True</td>
<td>Excludes gapped alignment columns meeting a threshold. Can accomodate
    reading frame. Returns modified Alignment.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>take_codon_positions</td>
<td>True</td>
<td>Extracts the specified codon position(s) from an alignment. 
    Returns an Alignment.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>take_named_seqs</td>
<td>True</td>
<td>Extracts (or everything but) named sequences. Returns a filtered
    sequences, alignment that satisified the condition, NotCompleted otherwise.</td>
<td>sequences, aligned, serialisable</td>
<td>sequences, aligned, serialisable</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>trim_stop_codons</td>
<td>True</td>
<td>Removes terminal stop codons. Returns sequences / alignment.</td>
<td>sequences, aligned, serialisable</td>
<td>sequences, aligned, serialisable</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.translate</td>
<td>select_translatable</td>
<td>True</td>
<td>Identifies most likely reading frame. Returns modified sequences / alignment,
    if it could be resolved, NotCompleted otherwise.</td>
<td>sequences, aligned</td>
<td>sequences</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.translate</td>
<td>translate_seqs</td>
<td>True</td>
<td>Translates sequences, assumes in correct reading frame.</td>
<td>sequences, aligned</td>
<td>sequences, aligned</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.tree</td>
<td>quick_tree</td>
<td>True</td>
<td>Neighbour Joining tree based on pairwise distances.</td>
<td>pairwise_distances</td>
<td>tree, serialisable</td>
<td>DistanceMatrix</td>
</tr>
<tr>
<td>cogent3.app.tree</td>
<td>scale_branches</td>
<td>True</td>
<td>Transforms tree branch lengths from nucleotide to codon, or the converse.
    Returns a Tree.</td>
<td>tree</td>
<td>tree, serialisable</td>
<td>PhyloNode, TreeNode</td>
</tr>
<tr>
<td>cogent3.app.tree</td>
<td>uniformize_tree</td>
<td>True</td>
<td>Standardises the orientation of unrooted trees. Returns a Tree.</td>
<td>tree</td>
<td>tree, serialisable</td>
<td>PhyloNode, TreeNode</td>
</tr>
<tr>
<td>cogent3.app.composable</td>
<td>user_function</td>
<td>True</td>
<td>wrapper class for user specified function</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>
37 rows x 7 columns</p></div></div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020, cogent3 Team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.1.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>