<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Apply a non-stationary nucleotide model to an alignment with 3 sequences &#8212; cogent3 2020.4 documentation</title>
    <link rel="stylesheet" href="../../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-dropdown.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/language_data.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../_static/require.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../../_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">
          cogent3</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../install.html">Install</a></li>
                <li><a href="../index.html">Docs</a></li>
                <li><a href="../draw/index.html">Gallery</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Project History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pycogent.html">How does <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> relate to PyCogent?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a><ul>
<li><a class="reference internal" href="#model-result"><code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li><a class="reference internal" href="#fitting-a-separate-nucleotide-model-to-each-codon-position">Fitting a separate nucleotide model to each codon position</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
              
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<div class="section" id="apply-a-non-stationary-nucleotide-model-to-an-alignment-with-3-sequences">
<h1>Apply a non-stationary nucleotide model to an alignment with 3 sequences<a class="headerlink" href="#apply-a-non-stationary-nucleotide-model-to-an-alignment-with-3-sequences" title="Permalink to this headline">¶</a></h1>
<p>We load some sample data first and select just 3 sequences, all via
<code class="docutils literal notranslate"><span class="pre">apps</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.app</span> <span class="kn">import</span> <span class="n">io</span><span class="p">,</span> <span class="n">sample</span>

<span class="n">reader</span> <span class="o">=</span> <span class="n">io</span><span class="o">.</span><span class="n">load_aligned</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">select_seqs</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">take_named_seqs</span><span class="p">(</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;Rhesus&quot;</span><span class="p">,</span> <span class="s2">&quot;Galago&quot;</span><span class="p">)</span>
<span class="n">process</span> <span class="o">=</span> <span class="n">reader</span> <span class="o">+</span> <span class="n">select_seqs</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">process</span><span class="p">(</span><span class="s2">&quot;data/primate_brca1.fasta&quot;</span><span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">names</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;Human&#39;, &#39;Rhesus&#39;, &#39;Galago&#39;]
</pre></div>
</div>
</div>
</div>
<p>We analyses these using the general Markov nucleotide, GN, model.
Because we analyse just 3 sequences, there is only one possible unrooted
tree. It’s not required to specify the tree in this instance.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.app</span> <span class="kn">import</span> <span class="n">evo</span>

<span class="n">gn</span> <span class="o">=</span> <span class="n">evo</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="s2">&quot;GN&quot;</span><span class="p">)</span>
<span class="n">gn</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>model(type=&#39;model&#39;, sm=&#39;GN&#39;, tree=None, name=None, sm_args=None, lf_args=None, time_het=None, param_rules=None, opt_args=None, split_codons=False, show_progress=False, verbose=False)
</pre></div>
</div>
</div>
</div>
<p>We apply this to <code class="docutils literal notranslate"><span class="pre">aln</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fitted</span> <span class="o">=</span> <span class="n">gn</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="nb">type</span><span class="p">(</span><span class="n">fitted</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>cogent3.app.result.model_result
</pre></div>
</div>
</div>
</div>
<div class="section" id="model-result">
<h2><code class="docutils literal notranslate"><span class="pre">model_result</span></code><a class="headerlink" href="#model-result" title="Permalink to this headline">¶</a></h2>
<p>As the output above indicates, <code class="docutils literal notranslate"><span class="pre">fitted</span></code> is a <code class="docutils literal notranslate"><span class="pre">model_result</span></code> object.</p>
<p>This object provides an interface for accessing attributes of a fitted model. The representation display (below), a styled table in a jupyter notebook, presents a summary view with the log-likelihood (<code class="docutils literal notranslate"><span class="pre">lnL</span></code>), number of free parameters (<code class="docutils literal notranslate"><span class="pre">nfp</span></code>) and whether all matrices satisfied the identifiability conditions diagonal largest in column (DLC) and a unique mapping of Q to P. (For description of these quantities and why they matter see <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/?term=8854662">Chang 1996</a> and <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/25503772">Kaehler et al</a>.)</p>
<p><code class="docutils literal notranslate"><span class="pre">model_result</span></code> has dictionary behaviour, hence the <code class="docutils literal notranslate"><span class="pre">key</span></code> column. This will be demonstrated below.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fitted</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<caption style="color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); align=top;"><span style="font-weight: bold;">GN</span><span></span></caption>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>key</th>
<th>lnL</th>
<th>nfp</th>
<th>DLC</th>
<th>unique_Q</th>
</thead>
<tbody>
<tr>
<td></td>
<td style="font-family: monospace !important;">-5964.0129</td>
<td style="font-family: monospace !important;">17</td>
<td>True</td>
<td>True</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>More detail on the fitted model are available via attributes. For instance, display the maximum likelihood estimates via the likelihood function attribute</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fitted</span><span class="o">.</span><span class="n">lf</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h4>Likelihood function statistics</h4>
<p>log-likelihood = -5964.0129</p>
<p>number of free parameters = 17</p>
<table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<caption style="color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); align=top;"><span style="font-weight: bold;">Global params</span><span></span></caption>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>A&gt;C</th>
<th>A&gt;G</th>
<th>A&gt;T</th>
<th>C&gt;A</th>
<th>C&gt;G</th>
<th>C&gt;T</th>
<th>G&gt;A</th>
<th>G&gt;C</th>
<th>G&gt;T</th>
<th>T&gt;A</th>
</thead>
<tbody>
<tr>
<td style="font-family: monospace !important;">1.0628</td>
<td style="font-family: monospace !important;">3.1833</td>
<td style="font-family: monospace !important;">1.0207</td>
<td style="font-family: monospace !important;">1.7952</td>
<td style="font-family: monospace !important;">2.3276</td>
<td style="font-family: monospace !important;">5.6847</td>
<td style="font-family: monospace !important;">9.0910</td>
<td style="font-family: monospace !important;">1.1136</td>
<td style="font-family: monospace !important;">0.8313</td>
<td style="font-family: monospace !important;">1.4997</td>
</tr>
</tbody>
</table>
<table>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>T&gt;C</th>
</thead>
<tbody>
<tr>
<td style="font-family: monospace !important;">3.5575</td>
</tr>
</tbody>
</table>

<table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<caption style="color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); align=top;"><span style="font-weight: bold;">Edge params</span><span></span></caption>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>edge</th>
<th>parent</th>
<th>length</th>
</thead>
<tbody>
<tr>
<td>Human</td>
<td>root</td>
<td style="font-family: monospace !important;">0.0214</td>
</tr>
<tr>
<td>Rhesus</td>
<td>root</td>
<td style="font-family: monospace !important;">0.0208</td>
</tr>
<tr>
<td>Galago</td>
<td>root</td>
<td style="font-family: monospace !important;">0.1781</td>
</tr>
</tbody>
</table>

<table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<caption style="color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); align=top;"><span style="font-weight: bold;">Motif params</span><span></span></caption>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>A</th>
<th>C</th>
<th>G</th>
<th>T</th>
</thead>
<tbody>
<tr>
<td style="font-family: monospace !important;">0.3740</td>
<td style="font-family: monospace !important;">0.1755</td>
<td style="font-family: monospace !important;">0.2098</td>
<td style="font-family: monospace !important;">0.2408</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fitted</span><span class="o">.</span><span class="n">lnL</span><span class="p">,</span> <span class="n">fitted</span><span class="o">.</span><span class="n">nfp</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>(-5964.012906757204, 17)
</pre></div>
</div>
</div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fitted</span><span class="o">.</span><span class="n">source</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>&#39;data/primate_brca1.fasta&#39;
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">model_result.tree</span></code> attribute is an “annotated tree”. Maximum likelihood estimates from the model have been assigned to the tree. Of particular significance, the “length” attribute corresponds to the expected number of substitutions (or ENS). For a non-stationary model, like GN, this can be different to the conventional length (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/25503772">Kaehler et al</a>).</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fitted</span><span class="o">.</span><span class="n">tree</span><span class="p">,</span> <span class="n">fitted</span><span class="o">.</span><span class="n">alignment</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>(Tree(&quot;(Human,Rhesus,Galago)root;&quot;),
 3 x 2814 dna alignment: Human[TGTGGCACAAA...], Rhesus[TGTGGCACAAA...], Galago[TGTGGCAAAAA...])
</pre></div>
</div>
</div>
</div>
<p>We can access the sum of all branch lengths. Either as “ENS” or “paralinear” using the <code class="docutils literal notranslate"><span class="pre">total_length()</span></code> method.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fitted</span><span class="o">.</span><span class="n">total_length</span><span class="p">(</span><span class="n">length_as</span><span class="o">=</span><span class="s2">&quot;paralinear&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>0.9280297233349861
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="fitting-a-separate-nucleotide-model-to-each-codon-position">
<h2>Fitting a separate nucleotide model to each codon position<a class="headerlink" href="#fitting-a-separate-nucleotide-model-to-each-codon-position" title="Permalink to this headline">¶</a></h2>
<p>Controlled by setting <code class="docutils literal notranslate"><span class="pre">split_codons=True</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gn</span> <span class="o">=</span> <span class="n">evo</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="s2">&quot;GN&quot;</span><span class="p">,</span> <span class="n">split_codons</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">fitted</span> <span class="o">=</span> <span class="n">gn</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="n">fitted</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<caption style="color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); align=top;"><span style="font-weight: bold;">GN</span><span></span></caption>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>key</th>
<th>lnL</th>
<th>nfp</th>
<th>DLC</th>
<th>unique_Q</th>
</thead>
<tbody>
<tr>
<td></td>
<td style="font-family: monospace !important;">-5866.9371</td>
<td style="font-family: monospace !important;">51</td>
<td>True</td>
<td>True</td>
</tr>
<tr>
<td style="font-family: monospace !important;">1</td>
<td style="font-family: monospace !important;">-1955.5141</td>
<td style="font-family: monospace !important;">17</td>
<td></td>
<td></td>
</tr>
<tr>
<td style="font-family: monospace !important;">2</td>
<td style="font-family: monospace !important;">-1934.2598</td>
<td style="font-family: monospace !important;">17</td>
<td></td>
<td></td>
</tr>
<tr>
<td style="font-family: monospace !important;">3</td>
<td style="font-family: monospace !important;">-1977.1632</td>
<td style="font-family: monospace !important;">17</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>The model fit statistics, <code class="docutils literal notranslate"><span class="pre">lnL</span></code> and <code class="docutils literal notranslate"><span class="pre">nfp</span></code> are now sums of the corresponding values from the fits to the individual positions. The <code class="docutils literal notranslate"><span class="pre">DLC</span></code> and <code class="docutils literal notranslate"><span class="pre">unique_Q</span></code> are also a summary across all models. These only achieve the value <code class="docutils literal notranslate"><span class="pre">True</span></code> when all matrices, from all models, satisfy the condition.</p>
<p>We get access to the likelihood functions of the individual positions via the indicated dict keys.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fitted</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h4>LF id: 3</h4>
<p>log-likelihood = -1977.1632</p>
<p>number of free parameters = 17</p>
<table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<caption style="color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); align=top;"><span style="font-weight: bold;">Global params</span><span></span></caption>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>A&gt;C</th>
<th>A&gt;G</th>
<th>A&gt;T</th>
<th>C&gt;A</th>
<th>C&gt;G</th>
<th>C&gt;T</th>
<th>G&gt;A</th>
<th>G&gt;C</th>
<th>G&gt;T</th>
</thead>
<tbody>
<tr>
<td style="font-family: monospace !important;">1.7484</td>
<td style="font-family: monospace !important;">3.4215</td>
<td style="font-family: monospace !important;">2.4111</td>
<td style="font-family: monospace !important;">2.3729</td>
<td style="font-family: monospace !important;">3.4060</td>
<td style="font-family: monospace !important;">16.1603</td>
<td style="font-family: monospace !important;">16.9889</td>
<td style="font-family: monospace !important;">2.0343</td>
<td style="font-family: monospace !important;">1.7432</td>
</tr>
</tbody>
</table>
<table>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>T&gt;A</th>
<th>T&gt;C</th>
</thead>
<tbody>
<tr>
<td style="font-family: monospace !important;">1.9241</td>
<td style="font-family: monospace !important;">5.2795</td>
</tr>
</tbody>
</table>

<table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<caption style="color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); align=top;"><span style="font-weight: bold;">Edge params</span><span></span></caption>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>edge</th>
<th>parent</th>
<th>length</th>
</thead>
<tbody>
<tr>
<td>Human</td>
<td>root</td>
<td style="font-family: monospace !important;">0.0243</td>
</tr>
<tr>
<td>Rhesus</td>
<td>root</td>
<td style="font-family: monospace !important;">0.0319</td>
</tr>
<tr>
<td>Galago</td>
<td>root</td>
<td style="font-family: monospace !important;">0.1797</td>
</tr>
</tbody>
</table>

<table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<caption style="color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); align=top;"><span style="font-weight: bold;">Motif params</span><span></span></caption>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>A</th>
<th>C</th>
<th>G</th>
<th>T</th>
</thead>
<tbody>
<tr>
<td style="font-family: monospace !important;">0.3375</td>
<td style="font-family: monospace !important;">0.1395</td>
<td style="font-family: monospace !important;">0.1647</td>
<td style="font-family: monospace !important;">0.3583</td>
</tr>
</tbody>
</table>
</div></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020, cogent3 Team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.1.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>