<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />

<!-- IE Compatibility tag -->
<meta http-equiv="X-UA-Compatible" content="IE=edge">

<title>cogent3 2020.4 documentation</title>

<link href="../../../.././_static/favicon.ico" rel="shortcut icon" type="image/x-icon"/>


<!-- Metadata -->
<meta name="description"
      content="cogent3 documentation site">



<!-- Facebook OpenGraph -->
<meta property="og:type" content="website"/>
<meta property="og:site" content="cogent3 documentation site"/>
<meta property="og:title" content="cogent3"/>
<meta property="og:description"
      content="Check the documentation site for cogent3"/>



<!-- HTML5 Shiv. For IE6-8 support of HTML5 elements -->
<!--[if lt IE 9]>
<script src="../../../.././_static/lib/html5shiv/html5shiv.min.js"></script>
<![endif]-->
<!-- CSS -->
<link rel="stylesheet" href="../../../../_static/jupyter-sphinx.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/thebelab.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/sphinx-dropdown.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/panels-bootstrap.min.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/gallery.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/gallery-binder.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/gallery-dataframe.css" type="text/css" />

<link rel="stylesheet" href="../../../.././_static/lib/bootswatch/litera/bootstrap.min.css">
<link rel="stylesheet" href="../../../.././_static/lib/font-awesome/css/font-awesome.min.css">
<link rel="stylesheet" href="../../../.././_static/lib/highlight/styles/default.css">

<!-- Docs Template style -->
<link rel="stylesheet" href="../../../.././_static/pygments.css">
<link rel="stylesheet" href="../../../.././_static/css/common.css">
<link rel="stylesheet" href="../../../.././_static/css/link.css">
<link rel="stylesheet" href="../../../.././_static/css/custom.css">

<!-- JavaScript -->

<script type="text/javascript" id="documentation_options" data-url_root="../../../../" src="../../../../_static/documentation_options.js"></script>
<script src="../../../.././_static/lib/jquery/jquery.min.js"></script>
<script src="../../../.././_static/lib/bootstrap/js/bootstrap.min.js"></script>
<script src="../../../.././_static/lib/highlight/highlight.pack.js"></script>
<script src="../../../.././_static/js/initializeHighlight.js"></script>
<script src="../../../../_static/jquery.js"></script>
<script src="../../../../_static/underscore.js"></script>
<script src="../../../../_static/doctools.js"></script>
<script src="../../../../_static/language_data.js"></script>
<script src="../../../../_static/thebelab-helper.js"></script>
<script src="../../../../_static/plotly-latest.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
<script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../../../_static/require.js"></script>
<script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
<script src="../../../.././_static/js/custom.js"></script>

</head>
<body class="d-flex flex-column">
<header>
    <nav id="navbar-main" class="navbar navbar-expand navbar-light bg-light">
    <a class="navbar-brand" href="../../../.././index.html">cogent3</a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-main-menu" aria-controls="navbar-main" aria-expanded="false" aria-label="Toggle navigation">
       <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar-main-menu" class="collapse navbar-collapse">
        <small class="navbar-text"><span id="navbar-version"></span></time></small>
        <div class="ml-auto d-flex align-items-center row">
            
            
            <div class="col-sm">
                <ul class="navbar-nav">
                    
                    
                    <li class="nav-item">
                        <a href="../../../install.html" class="nav-link" title="Install">Install</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../../../draw/index.html" class="nav-link" title="Gallery">Gallery</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../../../index.html" class="nav-link" title="Docs">Docs</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3" class="nav-link" title="GitHub">GitHub</a>
                    </li>
                    
                    
                    
                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" id="examplesDropdown" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Site</a>
                        <div class="dropdown-menu dropdown-menu-right" role="menu" aria-labelledby="examplesDropdown">
                            
                            <a href="../../../../general.html" class="dropdown-item" title="Posting Bugs">Posting Bugs</a>
                            
                            <a href="../../../../general.html" class="dropdown-item" title="Citation">Citation</a>
                            
                            <a href="../../../../general.html" class="dropdown-item" title="Support">Support</a>
                            
                            <a href="../../../../history.html" class="dropdown-item" title="History">History</a>
                            
                            <a href="../../../../projects.html" class="dropdown-item" title="Projects">Projects</a>
                            
                            <a href="../../../licenses.html" class="dropdown-item" title="License">License</a>
                            
                        </div>
                    </li>
                    
                    
                </ul>
            </div>
            
            <div role="search">
                <form id="Search-form" class="wy-form" action="../../../../search.html" method="get">
                    <input type="text" name="q" placeholder="Search docs" />
                    <input type="hidden" name="check_keywords" value="yes" />
                    <input type="hidden" name="area" value="default" />
                </form>
            </div>
        </div>
    </div>
</nav>
</header>

<nav class="side-nav">
    <div class="side-nav-container">
        
        <div class="side-nav-menu hidden">
        
            
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#minimal-installation">Minimal installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#install-with-graphing-tools">Install with graphing tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#install-with-developer-tools">Install with developer tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#installing-the-development-version">Installing the development version</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#installing-using-conda">Installing using <code class="docutils literal notranslate"><span class="pre">conda</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../install.html#manual-creation-of-the-conda-environment">Manual creation of the <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../../../index.html">Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#overview">Overview</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/warning.html">Warning – apps are alpha!</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app-overview.html">Overview of using apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#define-the-apps">Define the apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#using-apps-like-functions">Using apps like functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#composing-a-multi-step-process-from-several-apps">Composing a multi-step process from several apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#applying-a-process-to-multiple-data-records">Applying a process to multiple data records</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#other-important-features">Other important features</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#you-can-track-progress">You can track progress</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#you-can-do-parallel-computation">You can do parallel computation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#you-can-log-the-settings-and-data-analysed">You can log the settings and data analysed</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#all-of-the-above">All of the above</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/dstore.html">Specifying data for analysis</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#a-read-only-data-store">A read only data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#data-store-members">Data store “members”</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#showing-the-last-few-members">Showing the last few members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#filtering-a-data-store-for-specific-members">Filtering a data store for specific members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#looping-over-a-data-store">Looping over a data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#making-a-writeable-data-store">Making a writeable data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#tinydb-data-stores-are-special">TinyDB data stores are special</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/dstore.html#interrogating-run-logs">Interrogating run logs</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/not-completed.html">Tracking records that could not be processed</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#the-notcompleted-object">The <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#notcompleted-false-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> FALSE type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#notcompleted-error-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> ERROR type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#composed-functions-propagate-notcompleted-results">Composed functions propagate <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> results</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/available-apps.html">What apps are there?</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-progressive-align-app">The <code class="docutils literal notranslate"><span class="pre">progressive_align</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/align-nucleotide.html">Using a nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#nucleotide-alignment-with-default-settings">Nucleotide alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#specifying-the-substitution-model">Specifying the substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/align-codon.html">Using a codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#codon-alignment-with-default-settings">Codon alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#specifying-the-gap-parameters">Specifying the gap parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#specifying-the-substitution-model-and-parameters">Specifying the substitution model and parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/align-protein.html">Using a protein model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-protein.html#protein-alignment-with-default-settings">Protein alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-protein.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-protein.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-model-app">The <code class="docutils literal notranslate"><span class="pre">model</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model.html#model-result"><code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model.html#fitting-a-separate-nucleotide-model-to-each-codon-position">Fitting a separate nucleotide model to each codon position</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-with-tree.html#specify-the-tree-via-a-tree-instance">Specify the tree via a tree instance</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-with-tree.html#specify-the-tree-via-a-path">Specify the tree via a path.</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-with-tree.html#apply-the-model-to-an-alignment">Apply the model to an alignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-timehet.html#specify-a-distinct-rate-matrix-for-the-great-apes">Specify a distinct rate matrix for the great apes</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#create-and-apply-tabulate-stats-app">Create and apply <code class="docutils literal notranslate"><span class="pre">tabulate_stats</span></code> app</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#edge-parameters">Edge parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#global-parameters">Global parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#motif-parameters">Motif parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-ancestral-states.html">Reconstructing ancestral states</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-ancestral-states.html#define-the-ancestral-states-app">Define the <code class="docutils literal notranslate"><span class="pre">ancestral_states</span></code> app</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html#get-a-tree-with-branch-lengths-as-paralinear">Get a tree with branch lengths as paralinear</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html#getting-parameter-estimates">Getting parameter estimates</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-tr-nuc-model.html#using-the-gtr-model">Using the GTR model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-nstat-codon-model.html#we-can-obtain-the-tree-with-branch-lengths-as-ens">We can obtain the tree with branch lengths as ENS</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-tr-codon-model.html">Applying a time-reversible codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-tr-codon-model.html#using-the-conditional-nucleotide-form-codon-model">Using the conditional nucleotide form codon model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-hypothesis-app">The <code class="docutils literal notranslate"><span class="pre">hypothesis</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-hypothesis.html#hypothesis-result-has-attributes-and-keys"><code class="docutils literal notranslate"><span class="pre">hypothesis_result</span></code> has attributes and keys</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/evo-hypothesis.html#accessing-the-test-statistics">Accessing the test statistics</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/evo-hypothesis.html#the-null-hypothesis">The null hypothesis</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/evo-hypothesis.html#the-alternate-hypothesis">The alternate hypothesis</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-hypothesis.html#saving-hypothesis-results">Saving hypothesis results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-natsel-apps">The <code class="docutils literal notranslate"><span class="pre">natsel</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-natsel_sitehet.html#getting-the-individual-site-posterior-probabilities">Getting the individual site posterior probabilities</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-natsel_zhang.html#getting-the-posterior-probabilities-of-site-class-membership">Getting the posterior probabilities of site-class membership</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-natsel_zhang.html#getting-all-the-statistics-in-tabular-form">Getting all the statistics in tabular form</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-user-function-apps">The <code class="docutils literal notranslate"><span class="pre">user_function</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/user_function.html">Custom composable apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/user_function.html#a-simple-example">A simple example</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/user_function.html#renaming-sequences">Renaming sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/user_function.html#a-user-function-for-with-a-different-output-type">A user function for with a different output type</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/index.html">Cookbook</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html">Manipulating biological data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html#sequences-and-alignments">Sequences and Alignments</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html">Loading an alignment from a file</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-aligned-sequences">Loading aligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-unaligned-sequences">Loading unaligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#specifying-the-file-format">Specifying the file format</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#specifying-the-sequence-molecular-type">Specifying the sequence molecular type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#making-an-alignment-from-standard-python-objects">Making an alignment from standard python objects</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#from-a-series-of-strings">From a series of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#from-a-dict-of-strings">From a dict of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#stripping-label-characters-on-loading">Stripping label characters on loading</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-sequences-using-format-parsers">Loading sequences using format parsers</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-fasta-sequences-from-an-open-file-or-list-of-lines">Loading FASTA sequences from an open file or list of lines</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#handling-overloaded-fasta-sequence-labels">Handling overloaded FASTA sequence labels</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/alignments.html">Sequence Collections and Alignments</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#what-s-the-difference-between-alignment-and-arrayalignment">What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#basic-collection-objects">Basic Collection objects</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#adding-new-sequences-to-an-existing-collection-or-alignment">Adding new sequences to an existing collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#the-elements-of-a-collection-or-alignment">The elements of a collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#alignments">Alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#slicing-an-alignment">Slicing an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#computing-alignment-statistics">Computing alignment statistics</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html"><code class="docutils literal notranslate"><span class="pre">Sequence</span></code></a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html#dna-and-rna-sequences">DNA and RNA sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/protein_sequences.html">Protein sequences</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#creating-a-proteinsequence-with-a-name">Creating a ProteinSequence with a name</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#converting-a-dna-sequence-string-to-protein-sequence-string">Converting a DNA sequence string to protein sequence string</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#loading-protein-sequences-from-a-phylip-file">Loading protein sequences from a Phylip file</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/annotations.html">Annotations</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/annotations.html#annotations-with-coordinates">Annotations with coordinates</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#automated-introduction-from-reading-genbank-files">Automated introduction from reading genbank files</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#customising-annotation-construction-from-reading-a-genbank-file">Customising annotation construction from reading a genbank file</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#creating-directly-on-a-sequence">Creating directly on a sequence</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#via">Via</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#adding-as-a-series-or-item-wise">Adding as a series or item-wise</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#taking-the-union-of-annotations">Taking the union of annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#getting-annotation-coordinates">Getting annotation coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#annotations-have-shadows">Annotations have shadows</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#adding-to-a-sequence-member-of-an-alignment">Adding to a sequence member of an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#adding-to-an-alignment">Adding to an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#slicing-sequences-and-alignments-by-annotations">Slicing sequences and alignments by annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#slicing-by-pseudo-feature-or-feature-series">Slicing by pseudo-feature or feature series</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#slice-series-must-not-be-overlapping">Slice series must not be overlapping</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#you-can-slice-an-annotation-itself">You can slice an annotation itself</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#sequence-vs-alignment-slicing">Sequence vs Alignment slicing</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#copying-annotations">Copying annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#querying">Querying</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#querying-produces-objects-only-valid-for-their-source">Querying produces objects only valid for their source</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#querying-for-absent-annotation">Querying for absent annotation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#querying-features-that-span-gaps-in-alignments">Querying features that span gaps in alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#as-one-span-unifies-features-with-discontinuous-alignment-coordinates"><code class="docutils literal notranslate"><span class="pre">as_one_span</span></code> unifies features with discontinuous alignment coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#behaviour-of-annotations-on-nucleic-acid-sequences">Behaviour of annotations on nucleic acid sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#masking-annotated-regions">Masking annotated regions</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#masking-annotated-regions-on-alignments">Masking annotated regions on alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#you-can-take-mask-of-the-shadow">You can take mask of the shadow</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#what-features-of-a-certain-type-are-available">What features of a certain type are available?</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#getting-all-features-of-a-type-or-everything-but-that-type">Getting all features of a type, or everything but that type</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#getting-sequence-features-when-you-have-an-alignment-object">Getting sequence features when you have an alignment object</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#annotation-display-on-sequences">Annotation display on sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/moltypes.html">Molecular types</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#available-molecular-types">Available molecular types</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#getting-a-moltype">Getting a <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#using-a-moltype-to-get-ambiguity-codes">Using a <code class="docutils literal notranslate"><span class="pre">MolType</span></code> to get ambiguity codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#moltype-definition-of-degenerate-codes"><code class="docutils literal notranslate"><span class="pre">MolType</span></code> definition of degenerate codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#nucleic-acid-moltype-and-complementing">Nucleic acid <code class="docutils literal notranslate"><span class="pre">MolType</span></code> and complementing</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#making-sequences">Making sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#verify-sequences">Verify sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#making-a-custom-moltype">Making a custom <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/alphabet.html">Alphabets</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#alphabet-and-moltype"><code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> and <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#creating-tuple-alphabets">Creating tuple alphabets</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#convert-a-sequence-into-integers">Convert a sequence into integers</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#convert-integers-to-a-sequence">Convert integers to a sequence</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html#genetic-code">Genetic code</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/what_codes.html">Available genetic codes</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/what_codes.html#getting-a-genetic-code-with-get-code">Getting a genetic code with <code class="docutils literal notranslate"><span class="pre">get_code()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html">Translate DNA sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#translate-all-six-frames">Translate all six frames</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#find-out-how-many-stops-in-a-frame">Find out how many stops in a frame</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#translate-a-codon">Translate a codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#look-up-the-amino-acid-corresponding-to-a-single-codon">Look up the amino acid corresponding to a single codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#get-all-the-codons-for-one-amino-acid">Get all the codons for one amino acid</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#get-all-the-codons-for-a-group-of-amino-acids">Get all the codons for a group of amino acids</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#converting-the-codonalphabet-to-codon-series">Converting the <code class="docutils literal notranslate"><span class="pre">CodonAlphabet</span></code> to codon series</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#obtaining-the-codons-from-a-dnasequence-object">Obtaining the codons from a <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#translating-a-dna-sequence-with-a-terminating-stop-codon">Translating a DNA sequence with a terminating stop codon</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#by-removing-the-trailing-stop-codon-first">By removing the trailing stop codon first</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#by-slicing-the-dnasequence-first">By slicing the <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> first</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html#trees">Trees</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/simple_trees.html">Trees</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#loading-a-tree-from-a-file-and-visualizing-it-with-ascii-art">Loading a tree from a file and visualizing it with <code class="docutils literal notranslate"><span class="pre">ascii_art()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#writing-a-tree-to-a-file">Writing a tree to a file</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-individual-nodes-of-a-tree-by-name">Getting the individual nodes of a tree by name</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-name-of-a-node-or-a-tree">Getting the name of a node (or a tree)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#the-object-type-of-a-tree-and-its-nodes-is-the-same">The object type of a tree and its nodes is the same</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#working-with-the-nodes-of-a-tree">Working with the nodes of a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-path-between-two-tips-or-edges-connecting-edges">Getting the path between two tips or edges (connecting edges)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-distance-between-two-nodes">Getting the distance between two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-last-common-ancestor-lca-for-two-nodes">Getting the last common ancestor (LCA) for two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-all-the-ancestors-for-a-node">Getting all the ancestors for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-all-the-children-for-a-node">Getting all the children for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-all-the-distances-for-a-tree">Getting all the distances for a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-two-nodes-that-are-farthest-apart">Getting the two nodes that are farthest apart</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#get-the-nodes-within-a-given-distance">Get the nodes within a given distance</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#rerooting-trees">Rerooting trees</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#tree-representations">Tree representations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#tree-traversal">Tree traversal</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#selecting-subtrees">Selecting subtrees</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#tree-manipulation-methods">Tree manipulation methods</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/tables.html">Tabular data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-a-csv-file">Loading a csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-delimited-specifying-the-format">Loading delimited specifying the format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-a-set-number-of-lines-from-a-file">Loading a set number of lines from a file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#selectively-loading-parts-of-a-big-file">Selectively loading parts of a big file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-only-some-columns">Loading only some columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#load-raw-data-as-a-list-of-lists-of-strings">Load raw data as a list of lists of strings</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#make-a-table-from-header-and-rows">Make a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#make-a-table-from-a-dict">Make a table from a <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#specify-the-column-order-when-creating-from-a-dict">Specify the column order when creating from a <code class="docutils literal notranslate"><span class="pre">dict</span></code>.</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-the-table-with-an-index">Create the table with an index</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-a-pandas-dataframe">Create a table from a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-header-and-rows">Create a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-dict">Create a table from dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-a-2d-dict">Create a table from a 2D dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-that-has-complex-python-objects-as-elements">Create a table that has complex python objects as elements</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-an-empty-table">Create an empty table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#adding-a-new-column">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#add-a-title-and-a-legend-to-a-table">Add a title and a legend to a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#iterating-over-table-rows">Iterating over table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#how-many-rows-are-there">How many rows are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#how-many-columns-are-there">How many columns are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#iterating-over-table-columns">Iterating over table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#table-slicing-using-column-names">Table slicing using column names</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#table-slicing-using-indices">Table slicing using indices</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#changing-displayed-numerical-precision">Changing displayed numerical precision</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#change-digits-or-column-spacing">Change digits or column spacing</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#wrapping-tables-for-display">Wrapping tables for display</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#display-the-top-of-a-table-using-head">Display the top of a table using <code class="docutils literal notranslate"><span class="pre">head()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#display-the-bottom-of-a-table-using-tail">Display the bottom of a table using <code class="docutils literal notranslate"><span class="pre">tail()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#display-random-rows-from-a-table">Display random rows from a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#change-the-number-of-rows-displayed-by-repr">Change the number of rows displayed by <code class="docutils literal notranslate"><span class="pre">repr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#changing-column-headings">Changing column headings</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#id1">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-new-column-from-existing-ones">Create a new column from existing ones</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-table-data-as-a-numpy-array">Get table data as a numpy array</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-column-as-a-list">Get a table column as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-multiple-table-columns-as-a-list">Get multiple table columns as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-the-table-as-a-row-oriented-dict">Get the table as a row oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-the-table-as-a-column-oriented-dict">Get the table as a column oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-the-table-as-a-pandas-dataframe">Get the table as a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-of-counts-as-a-contingency-table">Get a table of counts as a contingency table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#appending-tables">Appending tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#summing-a-single-column">Summing a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#summing-multiple-columns-or-rows-strictly-numerical-data">Summing multiple columns or rows - strictly numerical data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#summing-multiple-columns-or-rows-with-mixed-non-numeric-numeric-data">Summing multiple columns or rows with mixed non-numeric/numeric data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#filtering-table-rows">Filtering table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#filtering-table-columns">Filtering table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#standard-sorting">Standard sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#reverse-sorting">Reverse sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#sorting-involving-multiple-columns-one-reversed">Sorting involving multiple columns, one reversed</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-raw-data-for-a-single-column">Getting raw data for a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-raw-data-for-multiple-columns">Getting raw data for multiple columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-distinct-values">Getting distinct values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#counting-occurrences-of-values">Counting occurrences of values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#counting-unique-values">Counting unique values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#joining-or-merging-tables">Joining or merging tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#transpose-a-table">Transpose a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#specify-markdown-as-the-str-format">Specify markdown as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#specify-latex-as-the-str-format">Specify latex as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-markdown-formatted-string">Get a table as a markdown formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-latex-formatted-string">Get a table as a latex formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-restructured-text-csv-table">Get a table as a restructured text csv-table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-restructured-text-grid-table">Get a table as a restructured text grid table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-a-latex-format-table-with-to-string">Getting a latex format table with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-a-bedgraph-format-with-to-string">Getting a bedGraph format with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-a-table-as-html">Getting a table as html</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#what-formats-can-be-written">What formats can be written?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#writing-a-latex-formmated-file">Writing a latex formmated file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#writing-delimited-formats">Writing delimited formats</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/building_alignments.html">Building alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_alignments.html#using-a-cogent3-progressive-aligner-for-nucleotides">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for nucleotides</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_alignments.html#using-a-cogent3-progressive-aligner-for-codons">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for codons</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_alignments.html#converting-gaps-from-aa-seq-alignment-to-nuc-seq-alignment">Converting gaps from aa-seq alignment to nuc seq alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html">Genetic distance calculation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#fast-pairwise-distance-estimation">Fast pairwise distance estimation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#computing-genetic-distances-using-the-alignment-object">Computing genetic distances using the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#using-the-distance-calculator-directly">Using the distance calculator directly</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#likelihood-based-pairwise-distance-estimation">Likelihood based pairwise distance estimation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/building_phylogenies.html">Building phylogenies</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#building-a-phylogenetic-tree-from-pairwise-distances">Building A Phylogenetic Tree From Pairwise Distances</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#directly-via-alignment-quick-tree">Directly via <code class="docutils literal notranslate"><span class="pre">alignment.quick_tree()</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#using-the-distancematrix-object">Using the <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#explicitly-via-distancematrix-and-cogent3-phylo-nj-nj">Explicitly via <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">cogent3.phylo.nj.nj()`</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#directly-from-a-pairwise-distance-dict">Directly from a pairwise distance <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#by-least-squares">By Least-squares</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#by-ml">By ML</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/evo_modelling.html">Evolutionary Analysis Using Likelihood</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-substitution-models">Specifying substitution models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#the-available-pre-defined-substitution-models">The available pre-defined substitution models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#getting-a-substitution-model-with-get-model">Getting a substitution model with <code class="docutils literal notranslate"><span class="pre">get_model()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#rate-heterogeneity-models">Rate heterogeneity models</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#for-nucleotide">For nucleotide</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#for-codon">For codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#for-protein">For protein</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/evo_modelling.html#making-a-likelihood-function">Making a likelihood function</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#providing-an-alignment-to-a-likelihood-function">Providing an alignment to a likelihood function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#scoping-parameters-on-trees-time-heterogeneous-models">Scoping parameters on trees – time heterogeneous models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-a-parameter-as-constant">Specifying a parameter as constant</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#providing-a-starting-value-for-a-parameter">Providing a starting value for a parameter</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#setting-parameter-bounds-for-optimisation">Setting parameter bounds for optimisation</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#setting-an-upper-bound-for-branch-length">Setting an upper bound for branch length</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-rate-heterogeneity-functions">Specifying rate heterogeneity functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-phylo-hmms">Specifying Phylo-HMMs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#fitting-likelihood-functions-choice-of-optimisers">Fitting likelihood functions - Choice of optimisers</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#how-to-check-your-optimisation-was-successful">How to check your optimisation was successful</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#overview-of-the-fitted-likelihood-function">Overview of the fitted likelihood function</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#log-likelihood-and-number-of-free-parameters">Log likelihood and number of free parameters</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#aikake-information-criterion">Aikake Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#bayesian-information-criterion">Bayesian Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#getting-maximum-likelihood-estimates">Getting maximum likelihood estimates</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#testing-hypotheses-using-likelihood-ratio-tests">Testing Hypotheses - Using Likelihood Ratio Tests</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#testing-hypotheses-by-parametric-bootstrapping">Testing Hypotheses - By parametric bootstrapping</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#determining-confidence-intervals-on-mles">Determining confidence intervals on MLEs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#saving-results">Saving results</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#reconstructing-ancestral-sequences">Reconstructing ancestral sequences</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/useful_utilities.html">Useful Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/useful_utilities.html#uniondict-a-dict-with-set-like-operations-and-keys-as-attributes"><code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> – a dict with set like operations and keys as attributes</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#accessing-elements-of-a-uniondict">Accessing elements of a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#updating-a-uniondict">Updating a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#accessing-a-non-existent-uniondict-key">Accessing a non-existent <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> key</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/useful_utilities.html#using-cogent3-s-optimisers-for-your-own-functions">Using Cogent3’s optimisers for your own functions</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/useful_utilities.html#miscellaneous-functions">Miscellaneous functions</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#force-a-variable-to-be-iterable">Force a variable to be iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#curry-a-function">Curry a function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#test-to-see-if-an-object-is-iterable">Test to see if an object is iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#test-to-see-if-an-object-is-a-single-char">Test to see if an object is a single char</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#flatten-a-deeply-nested-iterable">Flatten a deeply nested iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#test-to-determine-if-list-of-tuple">Test to determine if <code class="docutils literal notranslate"><span class="pre">list</span></code> of <code class="docutils literal notranslate"><span class="pre">tuple</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#create-a-case-insensitive-iterable">Create a case-insensitive iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#construct-a-distance-matrix-lookup-function">Construct a distance matrix lookup function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#check-class-types">Check class types</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#delegate-to-a-separate-object">Delegate to a separate object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#wrap-a-function-to-hide-from-a-class">Wrap a function to hide from a class</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#construct-a-constrained-container">Construct a constrained container</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/index.html">Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#data-manipulation">Data manipulation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/seq_features.html">Advanced sequence handling</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/complete_seq_features.html">Annotating alignments and sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/manipulating_tree_nodes.html#an-example-of-conditional-tree-node-modifications">An Example of Conditional Tree Node Modifications</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#general-data-analysis">General data analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/period_estimation.html">Estimating periodic signals</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/period_estimation.html#estimating-the-periods-in-a-signal">Estimating the periods in a signal</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#for-numerical-continuous-data">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#for-symbolic-data">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/period_estimation.html#estimating-power-for-specified-period">Estimating power for specified period</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id1">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id2">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/period_estimation.html#measuring-statistical-significance-of-periodic-signals">Measuring statistical significance of periodic signals</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id3">For numerical (continuous data)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id4">For symbolic data</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#modelling-evolution">Modelling Evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/simple.html">The simplest script</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#specifying-a-clade">Specifying a clade</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#specifying-a-stem">Specifying a stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#specifying-clade-and-stem">Specifying clade and stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#alternate-arguments-for-specifying-edges">Alternate arguments for specifying edges</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#applications-of-scoped-parameters">Applications of scoped parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/codon_models.html">Using codon models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#constructing-the-codon-substitution-model">Constructing the codon substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#standard-test-of-neutrality">Standard test of neutrality</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#rate-heterogeneity-model-variants">Rate-heterogeneity model variants</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#mixing-branch-and-site-heterogeneity">Mixing branch and site-heterogeneity</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#model-ho-no-rate-heterogeneity">Model Ho: no rate heterogeneity</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#model-ha-1-two-classes-of-gamma-distributed-but-independent-sites">Model Ha(1): two classes of gamma distributed but independent sites</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated">Model Ha(2): fast and slowly evolving sites are auto-correlated</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#a-model-with-patches-of-kappa">A model with patches of <code class="docutils literal notranslate"><span class="pre">kappa</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#phylogenetic-reconstruction">Phylogenetic Reconstruction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#look-for-the-single-best-tree">Look for the single best tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#a-more-rigorous-tree-space-search">A more rigorous tree space search</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#assessing-the-fit-for-a-pre-specified-tree-topology">Assessing the fit for a pre-specified tree topology</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#using-maximum-likelihood-for-measuring-tree-fit">Using maximum likelihood for measuring tree fit</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2 current"><a class="reference internal" href="../../index.html">API</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../index.html#utility-functions-for">Utility Functions For</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#loading-data-from-file">Loading data from file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#making-cogent3-types-from-standard-python-types">Making cogent3 types from standard Python types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#getting-commonly-used-cogent3-types">Getting commonly used cogent3 types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#displaying-cogent3-builtins">Displaying cogent3 builtins</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3 current"><a class="reference internal" href="../../index.html#the-major-cogent3-data-types">The Major cogent3 Data Types</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="../alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><ul class="current">
<li class="toctree-l5 current"><a class="current reference internal" href="#">SequenceCollection</a></li>
<li class="toctree-l5"><a class="reference internal" href="cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l5"><a class="reference internal" href="cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../table/classes/cogent3.util.table.Columns.html">Columns</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../draw/index.html">Image Gallery</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../draw/index.html#alignments-sequences">Alignments &amp; Sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html">Evaluating coevolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html#display-coevolution-as-a-heatmap">Display coevolution as a heatmap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-violin-plot">Display coevolution scores as a Violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-boxplot">Display coevolution scores as a Boxplot</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html">Dotplot</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#modifying-the-matching-parameters">Modifying the matching parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#displaying-dotplot-for-the-reverse-complement">Displaying dotplot for the reverse complement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#setting-plot-attributes">Setting plot attributes</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#all-options">All options</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-2.html">Dotplot with annotated sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-2.html#reloading-from-json">Reloading from json</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-2.html#removing-annotation-tracks">Removing annotation tracks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html">Counting gaps per sequence</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#plotting-counts-of-unique-gaps">Plotting counts of unique gaps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-bar-chart">Displaying unique gaps as a bar chart</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-violin-plot">Displaying unique gaps as a violin plot</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-box-plot">Displaying unique gaps as a box plot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-info-plot.html">Information analysis of an alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-info-plot.html#illustrated-with-a-simple-example">Illustrated with a simple example</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../draw/aln/plot_aln-info-plot.html#on-a-sample-data-set">On a sample data set</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html">Draw sequence logos</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html#drawing-logo-for-a-tfbs">Drawing logo for a TFBS</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html#drawing-a-sequence-logo-from-a-multiple-sequence-alignment">Drawing a sequence logo from a multiple sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html#sequence-logo-of-protein-alignment">Sequence logo of protein alignment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../draw/index.html#phylogenetic-trees">Phylogenetic Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-angular.html">Display a Phylogenetic Tree with a Angular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-angular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-circular.html">Display a Phylogenetic Tree with a Circular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-circular.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-circular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-radial.html">Display a Phylogenetic Tree with a Radial Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-radial.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html">Display a Phylogenetic Tree with a Square Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html#changing-scale-bar-placement">Changing scale bar placement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-support.html">Display a Phylogenetic Tree Showing Bootstrap Support</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-support.html#change-the-placement-of-support-text">Change the placement of support text</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../history.html">Project History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../history.html#cogent3-no-longer-includes-module-x-what-do-i-do"><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> no longer includes module <code class="docutils literal notranslate"><span class="pre">x</span></code>, what do I do?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../licenses.html">License</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../licenses.html#cogent3-license">Cogent3 License</a></li>
</ul>
</li>
</ul>

        
        </div>
        
    </div>
</nav>

<section id="main-section" class="container flex-grow-1 has-nav">
    
  <div class="section" id="sequencecollection">
<h1>SequenceCollection<a class="headerlink" href="#sequencecollection" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt id="cogent3.core.alignment.SequenceCollection">
<em class="property">class </em><code class="sig-name descname">SequenceCollection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">names</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">moltype</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">info</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">conversion_f</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">is_array</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">force_same_data</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">remove_duplicate_names</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">label_to_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">suppress_named_seqs</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection" title="Permalink to this definition">¶</a></dt>
<dd><p>Container for unaligned sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.num_seqs" title="cogent3.core.alignment.SequenceCollection.num_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_seqs</span></code></a></dt><dd><p>Returns the number of sequences in the alignment.</p>
</dd>
<dt><strong>seqs</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.add_seqs" title="cogent3.core.alignment.SequenceCollection.add_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_seqs</span></code></a>(other[, before_name, after_name])</p></td>
<td><p>Returns new object of class self with sequences from other added.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.annotate_from_gff" title="cogent3.core.alignment.SequenceCollection.annotate_from_gff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_from_gff</span></code></a>(f)</p></td>
<td><p>Copies annotations from gff-format file to self.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.apply_pssm" title="cogent3.core.alignment.SequenceCollection.apply_pssm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">apply_pssm</span></code></a>([pssm, path, background, …])</p></td>
<td><p>scores sequences using the specified pssm</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.copy" title="cogent3.core.alignment.SequenceCollection.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Returns deep copy of self.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.copy_annotations" title="cogent3.core.alignment.SequenceCollection.copy_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_annotations</span></code></a>(unaligned)</p></td>
<td><p>Copies annotations from seqs in unaligned to self, matching by name.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.counts" title="cogent3.core.alignment.SequenceCollection.counts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts</span></code></a>([motif_length, include_ambiguity, …])</p></td>
<td><p>returns dict of counts of motifs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.counts_per_seq" title="cogent3.core.alignment.SequenceCollection.counts_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts_per_seq</span></code></a>([motif_length, …])</p></td>
<td><p>returns dict of counts of motifs per sequence</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.deepcopy" title="cogent3.core.alignment.SequenceCollection.deepcopy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deepcopy</span></code></a>([sliced])</p></td>
<td><p>Returns deep copy of self.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.degap" title="cogent3.core.alignment.SequenceCollection.degap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degap</span></code></a>(**kwargs)</p></td>
<td><p>Returns copy in which sequences have no gaps.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.dotplot" title="cogent3.core.alignment.SequenceCollection.dotplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dotplot</span></code></a>([name1, name2, window, threshold, …])</p></td>
<td><p>make a dotplot between specified sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.entropy_per_seq" title="cogent3.core.alignment.SequenceCollection.entropy_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">entropy_per_seq</span></code></a>([motif_length, …])</p></td>
<td><p>Returns the Shannon entropy per sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_ambiguous_positions" title="cogent3.core.alignment.SequenceCollection.get_ambiguous_positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_ambiguous_positions</span></code></a>()</p></td>
<td><p>Returns dict of seq:{position:char} for ambiguous chars.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_identical_sets" title="cogent3.core.alignment.SequenceCollection.get_identical_sets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_identical_sets</span></code></a>([mask_degen])</p></td>
<td><p>returns sets of names for sequences that are identical</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_lengths" title="cogent3.core.alignment.SequenceCollection.get_lengths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_lengths</span></code></a>([include_ambiguity, allow_gap])</p></td>
<td><p>returns {name: seq length, …}</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_motif_probs" title="cogent3.core.alignment.SequenceCollection.get_motif_probs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_motif_probs</span></code></a>([alphabet, …])</p></td>
<td><p>Return a dictionary of motif probs, calculated as the averaged frequency across sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_seq" title="cogent3.core.alignment.SequenceCollection.get_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_seq</span></code></a>(seqname)</p></td>
<td><p>Return a sequence object for the specified seqname.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_seq_indices" title="cogent3.core.alignment.SequenceCollection.get_seq_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_seq_indices</span></code></a>(f[, negate])</p></td>
<td><p>Returns list of keys of seqs where f(row) is True.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_similar" title="cogent3.core.alignment.SequenceCollection.get_similar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_similar</span></code></a>(target[, min_similarity, …])</p></td>
<td><p>Returns new Alignment containing sequences similar to target.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_translation" title="cogent3.core.alignment.SequenceCollection.get_translation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_translation</span></code></a>([gc, incomplete_ok])</p></td>
<td><p>translate from nucleic acid to protein</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.has_terminal_stops" title="cogent3.core.alignment.SequenceCollection.has_terminal_stops"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_terminal_stops</span></code></a>([gc, allow_partial])</p></td>
<td><p>Returns True if any sequence has a terminal stop codon.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.is_ragged" title="cogent3.core.alignment.SequenceCollection.is_ragged"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_ragged</span></code></a>()</p></td>
<td><p>Returns True if alignment has sequences of different lengths.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.iter_selected" title="cogent3.core.alignment.SequenceCollection.iter_selected"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_selected</span></code></a>([seq_order, pos_order])</p></td>
<td><p>Iterates over elements in the alignment.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.iter_seqs" title="cogent3.core.alignment.SequenceCollection.iter_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_seqs</span></code></a>([seq_order])</p></td>
<td><p>Iterates over values (sequences) in the alignment, in order.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.omit_gap_runs" title="cogent3.core.alignment.SequenceCollection.omit_gap_runs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">omit_gap_runs</span></code></a>([allowed_run])</p></td>
<td><p>Returns new alignment where all seqs have runs of gaps &lt;=allowed_run.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.omit_gap_seqs" title="cogent3.core.alignment.SequenceCollection.omit_gap_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code></a>([allowed_gap_frac])</p></td>
<td><p>Returns new alignment with seqs that have &lt;= allowed_gap_frac.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.pad_seqs" title="cogent3.core.alignment.SequenceCollection.pad_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pad_seqs</span></code></a>([pad_length])</p></td>
<td><p>Returns copy in which sequences are padded to same length.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.probs_per_seq" title="cogent3.core.alignment.SequenceCollection.probs_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">probs_per_seq</span></code></a>([motif_length, …])</p></td>
<td><p>return MotifFreqsArray per sequence</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.rc" title="cogent3.core.alignment.SequenceCollection.rc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rc</span></code></a>()</p></td>
<td><p>Returns the reverse complement alignment</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.rename_seqs" title="cogent3.core.alignment.SequenceCollection.rename_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rename_seqs</span></code></a>(renamer)</p></td>
<td><p>returns new instance with sequences renamed</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.reverse_complement" title="cogent3.core.alignment.SequenceCollection.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse_complement</span></code></a>()</p></td>
<td><p>Returns the reverse complement alignment.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.set_repr_policy" title="cogent3.core.alignment.SequenceCollection.set_repr_policy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_repr_policy</span></code></a>([num_seqs, num_pos, ref_name])</p></td>
<td><p>specify policy for repr(self)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.strand_symmetry" title="cogent3.core.alignment.SequenceCollection.strand_symmetry"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strand_symmetry</span></code></a>([motif_length])</p></td>
<td><p>returns dict of strand symmetry test results per seq</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.take_seqs" title="cogent3.core.alignment.SequenceCollection.take_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">take_seqs</span></code></a>(seqs[, negate])</p></td>
<td><p>Returns new Alignment containing only specified seqs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.take_seqs_if" title="cogent3.core.alignment.SequenceCollection.take_seqs_if"><code class="xref py py-obj docutils literal notranslate"><span class="pre">take_seqs_if</span></code></a>(f[, negate])</p></td>
<td><p>Returns new Alignment containing seqs where f(row) is True.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_dict" title="cogent3.core.alignment.SequenceCollection.to_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dict</span></code></a>()</p></td>
<td><p>Returns the alignment as dict of names -&gt; strings.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_dna" title="cogent3.core.alignment.SequenceCollection.to_dna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dna</span></code></a>()</p></td>
<td><p>returns copy of self as an alignment of DNA moltype seqs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_fasta" title="cogent3.core.alignment.SequenceCollection.to_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_fasta</span></code></a>()</p></td>
<td><p>Return alignment in Fasta format</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_json" title="cogent3.core.alignment.SequenceCollection.to_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_json</span></code></a>()</p></td>
<td><p>returns json formatted string</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_moltype" title="cogent3.core.alignment.SequenceCollection.to_moltype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_moltype</span></code></a>(moltype)</p></td>
<td><p>returns copy of self with moltype seqs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_nexus" title="cogent3.core.alignment.SequenceCollection.to_nexus"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_nexus</span></code></a>(seq_type[, interleave_len])</p></td>
<td><p>Return alignment in NEXUS format and mapping to sequence ids</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_phylip" title="cogent3.core.alignment.SequenceCollection.to_phylip"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_phylip</span></code></a>()</p></td>
<td><p>Return alignment in PHYLIP format and mapping to sequence ids</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_protein" title="cogent3.core.alignment.SequenceCollection.to_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_protein</span></code></a>()</p></td>
<td><p>returns copy of self as an alignment of PROTEIN moltype seqs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_rich_dict" title="cogent3.core.alignment.SequenceCollection.to_rich_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rich_dict</span></code></a>()</p></td>
<td><p>returns detailed content including info and moltype attributes</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_rna" title="cogent3.core.alignment.SequenceCollection.to_rna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rna</span></code></a>()</p></td>
<td><p>returns copy of self as an alignment of RNA moltype seqs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.trim_stop_codons" title="cogent3.core.alignment.SequenceCollection.trim_stop_codons"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trim_stop_codons</span></code></a>([gc, allow_partial])</p></td>
<td><p>Removes any terminal stop codons from the sequences</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.with_modified_termini" title="cogent3.core.alignment.SequenceCollection.with_modified_termini"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_modified_termini</span></code></a>()</p></td>
<td><p>Changes the termini to include termini char instead of gapmotif.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.write" title="cogent3.core.alignment.SequenceCollection.write"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write</span></code></a>([filename, format])</p></td>
<td><p>Write the alignment to a file, preserving order of sequences.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.add_seqs">
<code class="sig-name descname">add_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">before_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">after_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.add_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns new object of class self with sequences from other added.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>other</strong></dt><dd><p>same class as self or coerceable to that class</p>
</dd>
<dt><strong>before_name</strong><span class="classifier">str</span></dt><dd><p>which sequence is added</p>
</dd>
<dt><strong>after_name</strong><span class="classifier">str</span></dt><dd><p>which sequence is added</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If both before_name and after_name are specified, the seqs will be
inserted using before_name.</p>
<p>By default the sequence is appended to the end of the alignment,
this can be changed by using either before_name or after_name arguments.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.annotate_from_gff">
<code class="sig-name descname">annotate_from_gff</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">f</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.annotate_from_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Copies annotations from gff-format file to self.</p>
<p>Matches by name of sequence. This method accepts string path
or pathlib.Path or file-like object (e.g. StringIO)</p>
<p>Skips sequences in the file that are not in self.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.apply_pssm">
<code class="sig-name descname">apply_pssm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pssm</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">path</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">background</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">pseudocount</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">names</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ui</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.apply_pssm" title="Permalink to this definition">¶</a></dt>
<dd><p>scores sequences using the specified pssm</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>pssm</strong><span class="classifier">profile.PSSM</span></dt><dd><p>if not provided, will be loaded from path</p>
</dd>
<dt><strong>path</strong></dt><dd><p>path to either a jaspar or cisbp matrix (path must end have a suffix
matching the format).</p>
</dd>
<dt><strong>pseudocount</strong></dt><dd><p>adjustment for zero in matrix</p>
</dd>
<dt><strong>names</strong></dt><dd><p>returns only scores for these sequences and in the name order</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy array of log2 based scores at every position</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns deep copy of self.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.copy_annotations">
<code class="sig-name descname">copy_annotations</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">unaligned</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.copy_annotations" title="Permalink to this definition">¶</a></dt>
<dd><p>Copies annotations from seqs in unaligned to self, matching by name.</p>
<p>Alignment programs like ClustalW don’t preserve annotations,
so this method is available to copy annotations off the unaligned
sequences.</p>
<p>unaligned should be a dictionary of Sequence instances.</p>
<p>Ignores sequences that are not in self, so safe to use on larger dict
of seqs that are not in the current collection/alignment.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.counts">
<code class="sig-name descname">counts</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">motif_length</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">include_ambiguity</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">allow_gap</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">exclude_unobserved</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.counts" title="Permalink to this definition">¶</a></dt>
<dd><p>returns dict of counts of motifs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of elements per character.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gaps</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>only non-overlapping motifs are counted</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.counts_per_seq">
<code class="sig-name descname">counts_per_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">motif_length</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">include_ambiguity</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">allow_gap</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">exclude_unobserved</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.counts_per_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>returns dict of counts of motifs per sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of characters per tuple.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gap</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>only non-overlapping motifs are counted</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.deepcopy">
<code class="sig-name descname">deepcopy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sliced</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.deepcopy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns deep copy of self.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.degap">
<code class="sig-name descname">degap</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.degap" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns copy in which sequences have no gaps.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.dotplot">
<code class="sig-name descname">dotplot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name1</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name2</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">window</span><span class="o">=</span><span class="default_value">20</span></em>, <em class="sig-param"><span class="n">threshold</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_gap</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">width</span><span class="o">=</span><span class="default_value">500</span></em>, <em class="sig-param"><span class="n">title</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">rc</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">show_progress</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.dotplot" title="Permalink to this definition">¶</a></dt>
<dd><p>make a dotplot between specified sequences. Random sequences
chosen if names not provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name1, name2</strong><span class="classifier">str or None</span></dt><dd><p>names of sequences. If one is not provided, a random choice is made</p>
</dd>
<dt><strong>window</strong><span class="classifier">int</span></dt><dd><p>k-mer size for comparison between sequences</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">int</span></dt><dd><p>windows where the sequences are identical &gt;= threshold are a match</p>
</dd>
<dt><strong>min_gap</strong><span class="classifier">int</span></dt><dd><p>permitted gap for joining adjacent line segments, default is no gap
joining</p>
</dd>
<dt><strong>width</strong><span class="classifier">int</span></dt><dd><p>figure width. Figure height is computed based on the ratio of
len(seq1) / len(seq2)</p>
</dd>
<dt><strong>title</strong></dt><dd><p>title for the plot</p>
</dd>
<dt><strong>rc</strong><span class="classifier">bool or None</span></dt><dd><p>include dotplot of reverse compliment also. Only applies to Nucleic
acids moltypes</p>
</dd>
<dt><strong>Returns</strong></dt><dd></dd>
<dt><strong>——-</strong></dt><dd></dd>
<dt><strong>a Drawable or AnnotatedDrawable</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.entropy_per_seq">
<code class="sig-name descname">entropy_per_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">motif_length</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">include_ambiguity</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">allow_gap</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">exclude_unobserved</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">alert</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.entropy_per_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the Shannon entropy per sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length: int</strong></dt><dd><p>number of characters per tuple.</p>
</dd>
<dt><strong>include_ambiguity: bool</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gap: bool</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved: bool</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>For motif_length &gt; 1, it’s advisable to specify exclude_unobserved=True,
this avoids unnecessary calculations.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.get_ambiguous_positions">
<code class="sig-name descname">get_ambiguous_positions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_ambiguous_positions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dict of seq:{position:char} for ambiguous chars.</p>
<p>Used in likelihood calculations.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.get_identical_sets">
<code class="sig-name descname">get_identical_sets</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mask_degen</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_identical_sets" title="Permalink to this definition">¶</a></dt>
<dd><p>returns sets of names for sequences that are identical</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>mask_degen</strong></dt><dd><p>if True, degenerate characters are ignored</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.get_lengths">
<code class="sig-name descname">get_lengths</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_ambiguity</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">allow_gap</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>returns {name: seq length, …}</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gaps</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.get_motif_probs">
<code class="sig-name descname">get_motif_probs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">include_ambiguity</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">exclude_unobserved</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">allow_gap</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">pseudocount</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_motif_probs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary of motif probs, calculated as the averaged
frequency across sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_ambiguity</strong></dt><dd><p>if True resolved ambiguous codes are
included in estimation of frequencies, default is False.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, motifs that are not present in
the alignment are excluded from the returned dictionary,
default is False.</p>
</dd>
<dt><strong>allow_gap</strong></dt><dd><p>allow gap motif</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>only non-overlapping motifs are counted</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.get_seq">
<code class="sig-name descname">get_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqname</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a sequence object for the specified seqname.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.get_seq_indices">
<code class="sig-name descname">get_seq_indices</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">f</span></em>, <em class="sig-param"><span class="n">negate</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_seq_indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of keys of seqs where f(row) is True.</p>
<p>List will be in the same order as self.names, if present.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.get_similar">
<code class="sig-name descname">get_similar</code><span class="sig-paren">(</span><em class="sig-param">target</em>, <em class="sig-param">min_similarity=0.0</em>, <em class="sig-param">max_similarity=1.0</em>, <em class="sig-param">metric=&lt;cogent3.util.transform.for_seq object&gt;</em>, <em class="sig-param">transform=None</em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_similar" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns new Alignment containing sequences similar to target.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>target</strong></dt><dd><p>sequence object to compare to. Can be in the alignment.</p>
</dd>
<dt><strong>min_similarity</strong></dt><dd><p>minimum similarity that will be kept. Default 0.0.</p>
</dd>
<dt><strong>max_similarity</strong></dt><dd><p>maximum similarity that will be kept. Default 1.0.
(Note that both min_similarity and max_similarity are inclusive.)
metric
similarity function to use. Must be f(first_seq, second_seq).</p>
</dd>
<dt><strong>The default metric is fraction similarity, ranging from 0.0 (0%</strong></dt><dd></dd>
<dt><strong>identical) to 1.0 (100% identical). The Sequence classes have lots</strong></dt><dd></dd>
<dt><strong>of methods that can be passed in as unbound methods to act as the</strong></dt><dd></dd>
<dt><strong>metric, e.g. frac_same_gaps.</strong></dt><dd></dd>
<dt><strong>transform</strong></dt><dd><p>transformation function to use on the sequences before
the metric is calculated. If None, uses the whole sequences in each
case. A frequent transformation is a function that returns a specified
range of a sequence, e.g. eliminating the ends. Note that the
transform applies to both the real sequence and the target sequence.</p>
</dd>
<dt><strong>WARNING: if the transformation changes the type of the sequence (e.g.</strong></dt><dd></dd>
<dt><strong>extracting a string from an RnaSequence object), distance metrics that</strong></dt><dd></dd>
<dt><strong>depend on instance data of the original class may fail.</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.get_translation">
<code class="sig-name descname">get_translation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gc</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">incomplete_ok</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_translation" title="Permalink to this definition">¶</a></dt>
<dd><p>translate from nucleic acid to protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>genetic code, either the number or name
(use cogent3.core.genetic_code.available_codes)</p>
</dd>
<dt><strong>incomplete_ok</strong><span class="classifier">bool</span></dt><dd><p>codons that are mixes of nucleotide and gaps converted to ‘?’.
raises a ValueError if False</p>
</dd>
<dt><strong>kwargs</strong></dt><dd><p>related to construction of the resulting object</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>A new instance of self translated into protein</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.has_terminal_stops">
<code class="sig-name descname">has_terminal_stops</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gc</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">allow_partial</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.has_terminal_stops" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if any sequence has a terminal stop codon.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>genetic code object</p>
</dd>
<dt><strong>allow_partial</strong></dt><dd><p>if True and the sequence length is not divisible
by 3, ignores the 3’ terminal incomplete codon</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="cogent3.core.alignment.SequenceCollection.is_array">
<code class="sig-name descname">is_array</code><em class="property"> = {'array', 'array_seqs'}</em><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.is_array" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.is_ragged">
<code class="sig-name descname">is_ragged</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.is_ragged" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if alignment has sequences of different lengths.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.iter_selected">
<code class="sig-name descname">iter_selected</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq_order</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">pos_order</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.iter_selected" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates over elements in the alignment.</p>
<p>seq_order (names) can be used to select a subset of seqs.
pos_order (positions) can be used to select a subset of positions.</p>
<p>Always iterates along a seq first, then down a position (transposes
normal order of a[i][j]; possibly, this should change)..</p>
<p>WARNING: Alignment.iter_selected() is not the same as alignment.iteritems()
(which is the built-in dict iteritems that iterates over key-value
pairs).</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.iter_seqs">
<code class="sig-name descname">iter_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq_order</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.iter_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates over values (sequences) in the alignment, in order.</p>
<p>seq_order: list of keys giving the order in which seqs will be returned.
Defaults to self.Names. Note that only these sequences will be
returned, and that KeyError will be raised if there are sequences
in order that have been deleted from the Alignment. If self.Names
is None, returns the sequences in the same order as
self.named_seqs.values().</p>
<p>Use map(f, self.seqs()) to apply the constructor f to each seq. f must
accept a single list as an argument.</p>
<p>Always returns references to the same objects that are values of the
alignment.</p>
</dd></dl>

<dl class="py attribute">
<dt id="cogent3.core.alignment.SequenceCollection.moltype">
<code class="sig-name descname">moltype</code><em class="property"> = MolType(('\x00', '\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\x07', '\x08', '\t', '\n', '\x0b', '\x0c', '\r', '\x0e', '\x0f', '\x10', '\x11', '\x12', '\x13', '\x14', '\x15', '\x16', '\x17', '\x18', '\x19', '\x1a', '\x1b', '\x1c', '\x1d', '\x1e', '\x1f', ' ', '!', '&quot;', '#', '$', '%', '&amp;', &quot;'&quot;, '(', ')', '*', '+', ',', '-', '.', '/', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ':', ';', '&lt;', '=', '&gt;', '?', '&#64;', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\', ']', '^', '_', '`', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '~', '\x7f', '\x80', '\x81', '\x82', '\x83', '\x84', '\x85', '\x86', '\x87', '\x88', '\x89', '\x8a', '\x8b', '\x8c', '\x8d', '\x8e', '\x8f', '\x90', '\x91', '\x92', '\x93', '\x94', '\x95', '\x96', '\x97', '\x98', '\x99', '\x9a', '\x9b', '\x9c', '\x9d', '\x9e', '\x9f', '\xa0', '¡', '¢', '£', '¤', '¥', '¦', '§', '¨', '©', 'ª', '«', '¬', '\xad', '®', '¯', '°', '±', '²', '³', '´', 'µ', '¶', '·', '¸', '¹', 'º', '»', '¼', '½', '¾', '¿', 'À', 'Á', 'Â', 'Ã', 'Ä', 'Å', 'Æ', 'Ç', 'È', 'É', 'Ê', 'Ë', 'Ì', 'Í', 'Î', 'Ï', 'Ð', 'Ñ', 'Ò', 'Ó', 'Ô', 'Õ', 'Ö', '×', 'Ø', 'Ù', 'Ú', 'Û', 'Ü', 'Ý', 'Þ', 'ß', 'à', 'á', 'â', 'ã', 'ä', 'å', 'æ', 'ç', 'è', 'é', 'ê', 'ë', 'ì', 'í', 'î', 'ï', 'ð', 'ñ', 'ò', 'ó', 'ô', 'õ', 'ö', '÷', 'ø', 'ù', 'ú', 'û', 'ü', 'ý', 'þ', 'ÿ'))</em><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.moltype" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.num_seqs">
<em class="property">property </em><code class="sig-name descname">num_seqs</code><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.num_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of sequences in the alignment.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.omit_gap_runs">
<code class="sig-name descname">omit_gap_runs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">allowed_run</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.omit_gap_runs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns new alignment where all seqs have runs of gaps &lt;=allowed_run.</p>
<p>Note that seqs with exactly allowed_run gaps are not deleted.
Default is for allowed_run to be 1 (i.e. no consecutive gaps allowed).</p>
<p>Because the test for whether the current gap run exceeds the maximum
allowed gap run is only triggered when there is at least one gap, even
negative values for allowed_run will still let sequences with no gaps
through.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.omit_gap_seqs">
<code class="sig-name descname">omit_gap_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">allowed_gap_frac</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.omit_gap_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns new alignment with seqs that have &lt;= allowed_gap_frac.</p>
<p>allowed_gap_frac should be a fraction between 0 and 1 inclusive.
Default is 0.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.pad_seqs">
<code class="sig-name descname">pad_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pad_length</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.pad_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns copy in which sequences are padded to same length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>pad_length</strong></dt><dd><p>Length all sequences are to be padded to.  Will pad
to max sequence length if pad_length is None or less than max
length.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.probs_per_seq">
<code class="sig-name descname">probs_per_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">motif_length</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">include_ambiguity</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">allow_gap</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">exclude_unobserved</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">alert</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.probs_per_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>return MotifFreqsArray per sequence</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.rc">
<code class="sig-name descname">rc</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement alignment</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.rename_seqs">
<code class="sig-name descname">rename_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">renamer</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.rename_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>returns new instance with sequences renamed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>renamer</strong><span class="classifier">callable</span></dt><dd><p>function that will take current sequences and return the new one</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement alignment. A synonymn for rc.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.seqs">
<em class="property">property </em><code class="sig-name descname">seqs</code><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.seqs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.set_repr_policy">
<code class="sig-name descname">set_repr_policy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_seqs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">num_pos</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ref_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.set_repr_policy" title="Permalink to this definition">¶</a></dt>
<dd><p>specify policy for repr(self)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>num_seqs</strong><span class="classifier">int or None</span></dt><dd><p>number of sequences to include in represented display.</p>
</dd>
<dt><strong>num_pos</strong><span class="classifier">int or None</span></dt><dd><p>length of sequences to include in represented display.</p>
</dd>
<dt><strong>ref_name</strong><span class="classifier">str or None</span></dt><dd><p>name of sequence to be placed first, or “longest” (default).
If latter, indicates longest sequence will be chosen.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.strand_symmetry">
<code class="sig-name descname">strand_symmetry</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">motif_length</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.strand_symmetry" title="Permalink to this definition">¶</a></dt>
<dd><p>returns dict of strand symmetry test results per seq</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.take_seqs">
<code class="sig-name descname">take_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span></em>, <em class="sig-param"><span class="n">negate</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.take_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns new Alignment containing only specified seqs.</p>
<p>Note that the seqs in the new alignment will be references to the
same objects as the seqs in the old alignment.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.take_seqs_if">
<code class="sig-name descname">take_seqs_if</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">f</span></em>, <em class="sig-param"><span class="n">negate</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.take_seqs_if" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns new Alignment containing seqs where f(row) is True.</p>
<p>Note that the seqs in the new Alignment are the same objects as the
seqs in the old Alignment, not copies.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the alignment as dict of names -&gt; strings.</p>
<p>Note: returns strings, NOT Sequence objects.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_dna">
<code class="sig-name descname">to_dna</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_dna" title="Permalink to this definition">¶</a></dt>
<dd><p>returns copy of self as an alignment of DNA moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_fasta">
<code class="sig-name descname">to_fasta</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Return alignment in Fasta format</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>make_seqlabel</strong></dt><dd><p>callback function that takes the seq object and
returns a label str</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_json">
<code class="sig-name descname">to_json</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_json" title="Permalink to this definition">¶</a></dt>
<dd><p>returns json formatted string</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_moltype">
<code class="sig-name descname">to_moltype</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">moltype</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_moltype" title="Permalink to this definition">¶</a></dt>
<dd><p>returns copy of self with moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_nexus">
<code class="sig-name descname">to_nexus</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq_type</span></em>, <em class="sig-param"><span class="n">interleave_len</span><span class="o">=</span><span class="default_value">50</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_nexus" title="Permalink to this definition">¶</a></dt>
<dd><p>Return alignment in NEXUS format and mapping to sequence ids</p>
<dl class="simple">
<dt><strong>NOTE</strong> Not that every sequence in the alignment MUST come from</dt><dd><p>a different species!! (You can concatenate multiple sequences from
same species together before building tree)</p>
</dd>
</dl>
<p>seq_type: dna, rna, or protein</p>
<p>Raises exception if invalid alignment</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_phylip">
<code class="sig-name descname">to_phylip</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_phylip" title="Permalink to this definition">¶</a></dt>
<dd><p>Return alignment in PHYLIP format and mapping to sequence ids</p>
<p>raises exception if invalid alignment</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_protein">
<code class="sig-name descname">to_protein</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>returns copy of self as an alignment of PROTEIN moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_rich_dict">
<code class="sig-name descname">to_rich_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_rich_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>returns detailed content including info and moltype attributes</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.to_rna">
<code class="sig-name descname">to_rna</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_rna" title="Permalink to this definition">¶</a></dt>
<dd><p>returns copy of self as an alignment of RNA moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.trim_stop_codons">
<code class="sig-name descname">trim_stop_codons</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gc</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">allow_partial</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.trim_stop_codons" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes any terminal stop codons from the sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>genetic code object</p>
</dd>
<dt><strong>allow_partial</strong></dt><dd><p>if True and the sequence length is not divisible
by 3, ignores the 3’ terminal incomplete codon</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.with_modified_termini">
<code class="sig-name descname">with_modified_termini</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.with_modified_termini" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes the termini to include termini char instead of gapmotif.</p>
<p>Useful to correct the standard gap char output by most
alignment programs when aligned sequences have different ends.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.alignment.SequenceCollection.write">
<code class="sig-name descname">write</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">format</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Write the alignment to a file, preserving order of sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>filename</strong></dt><dd><p>name of the sequence file</p>
</dd>
<dt><strong>format</strong></dt><dd><p>format of the sequence file</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If format is None, will attempt to infer format from the filename
suffix.</p>
</dd></dl>

</dd></dl>

</div>


</section>

<footer class="footer w-100 pt-2 border-top bg-light">
    <div id="footer-nav" class="bg-light container">
        <div id="navbar-footer" class="row">
            
            <div class="col">
                <dl class="text-center mb-1">
                    
                    
                    
                </dl>
            </div>
            
        </div>
    </div>
    <div id="footer-info" class="bg-light pl-2 pb-1 w-100">
        <div><span class="fa fa-copyright"></span>2020, cogent3 Team
        <a id="author-link" href=""></a> </div>
        <div>Powered by <a href="http://sphinx-doc.org/">Sphinx</a> 3.2.1 and the <a href="https://github.com/Bernardo-MG/sphinx-docs-theme">Sphinx Docs Theme</a></div>
    </div>
</footer>

</body>
</html>