<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Sequence Collections and Alignments &#8212; cogent3 2020.4 documentation</title>
    <link rel="stylesheet" href="../../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-dropdown.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/language_data.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../_static/require.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../../_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">
          cogent3</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../install.html">Install</a></li>
                <li><a href="../index.html">Docs</a></li>
                <li><a href="../draw/index.html">Gallery</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Project History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pycogent.html">How does <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> relate to PyCogent?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Sequence Collections and Alignments</a><ul>
<li><a class="reference internal" href="#what-s-the-difference-between-alignment-and-arrayalignment">What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?</a></li>
<li><a class="reference internal" href="#basic-collection-objects">Basic Collection objects</a><ul>
<li><a class="reference internal" href="#constructing-a-sequencecollection-or-alignment-object-from-strings">Constructing a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from strings</a></li>
<li><a class="reference internal" href="#constructing-a-arrayalignment-using-make-aligned-seqs">Constructing a <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> using <code class="docutils literal notranslate"><span class="pre">make_aligned_seqs</span></code></a></li>
<li><a class="reference internal" href="#converting-a-sequencecollection-to-fasta-format">Converting a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> to FASTA format</a></li>
</ul>
</li>
<li><a class="reference internal" href="#adding-new-sequences-to-an-existing-collection-or-alignment">Adding new sequences to an existing collection or alignment</a><ul>
<li><a class="reference internal" href="#appending-the-sequences">Appending the sequences</a></li>
<li><a class="reference internal" href="#inserting-the-sequences">Inserting the sequences</a></li>
<li><a class="reference internal" href="#inserting-sequence-s-based-on-their-alignment-to-a-reference-sequence">Inserting sequence(s) based on their alignment to a reference sequence</a><ul>
<li><a class="reference internal" href="#removing-all-columns-with-gaps-in-a-named-sequence">Removing all columns with gaps in a named sequence</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#the-elements-of-a-collection-or-alignment">The elements of a collection or alignment</a><ul>
<li><a class="reference internal" href="#accessing-individual-sequences-from-a-collection-or-alignment-by-name">Accessing individual sequences from a collection or alignment by name</a></li>
<li><a class="reference internal" href="#slice-the-sequences-from-an-alignment-like-a-list">Slice the sequences from an alignment like a list</a></li>
<li><a class="reference internal" href="#getting-a-subset-of-sequences-from-the-alignment">Getting a subset of sequences from the alignment</a></li>
</ul>
</li>
<li><a class="reference internal" href="#alignments">Alignments</a><ul>
<li><a class="reference internal" href="#creating-an-alignment-object-from-a-sequencecollection">Creating an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code></a></li>
<li><a class="reference internal" href="#convert-alignment-to-dna-rna-or-protein-moltypes">Convert alignment to DNA, RNA or PROTEIN moltypes</a></li>
<li><a class="reference internal" href="#handling-gaps">Handling gaps</a><ul>
<li><a class="reference internal" href="#remove-all-gaps-from-an-alignment-in-fasta-format">Remove all gaps from an alignment in FASTA format</a></li>
</ul>
</li>
<li><a class="reference internal" href="#writing-sequences-to-file">Writing sequences to file</a></li>
<li><a class="reference internal" href="#converting-an-alignment-to-fasta-format">Converting an alignment to FASTA format</a></li>
<li><a class="reference internal" href="#converting-an-alignment-into-phylip-format">Converting an alignment into Phylip format</a></li>
<li><a class="reference internal" href="#converting-an-alignment-to-a-list-of-strings">Converting an alignment to a list of strings</a></li>
</ul>
</li>
<li><a class="reference internal" href="#slicing-an-alignment">Slicing an alignment</a><ul>
<li><a class="reference internal" href="#by-rows-sequences">By rows (sequences)</a></li>
<li><a class="reference internal" href="#getting-a-single-column-from-an-alignment">Getting a single column from an alignment</a></li>
<li><a class="reference internal" href="#getting-a-region-of-contiguous-columns">Getting a region of contiguous columns</a></li>
<li><a class="reference internal" href="#iterating-over-alignment-positions">Iterating over alignment positions</a></li>
<li><a class="reference internal" href="#getting-codon-3rd-positions-from-alignment">Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">Alignment</span></code></a></li>
<li><a class="reference internal" href="#getting-codon-3rd-positions-from-arrayalignment">Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code></a></li>
<li><a class="reference internal" href="#filtering-positions">Filtering positions</a><ul>
<li><a class="reference internal" href="#trim-terminal-stop-codons">Trim terminal stop codons</a></li>
<li><a class="reference internal" href="#eliminating-columns-with-non-nucleotide-characters">Eliminating columns with non-nucleotide characters</a></li>
<li><a class="reference internal" href="#getting-all-variable-positions-from-an-alignment">Getting all variable positions from an alignment</a></li>
<li><a class="reference internal" href="#getting-all-constant-positions-from-an-alignment">Getting all constant positions from an alignment</a></li>
<li><a class="reference internal" href="#getting-all-variable-codons-from-an-alignment">Getting all variable codons from an alignment</a></li>
</ul>
</li>
<li><a class="reference internal" href="#filtering-sequences">Filtering sequences</a><ul>
<li><a class="reference internal" href="#extracting-sequences-by-sequence-identifier-into-a-new-alignment-object">Extracting sequences by sequence identifier into a new alignment object</a></li>
<li><a class="reference internal" href="#extracting-sequences-using-an-arbitrary-function-into-a-new-alignment-object">Extracting sequences using an arbitrary function into a new alignment object</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#computing-alignment-statistics">Computing alignment statistics</a><ul>
<li><a class="reference internal" href="#getting-motif-counts">Getting motif counts</a></li>
<li><a class="reference internal" href="#computing-motif-probabilities-from-an-alignment">Computing motif probabilities from an alignment</a></li>
<li><a class="reference internal" href="#working-with-alignment-gaps">Working with alignment gaps</a><ul>
<li><a class="reference internal" href="#filtering-extracted-columns-for-the-gap-character">Filtering extracted columns for the gap character</a></li>
<li><a class="reference internal" href="#calculating-the-gap-fraction">Calculating the gap fraction</a></li>
<li><a class="reference internal" href="#extracting-maps-of-aligned-to-unaligned-positions-i-e-gap-maps">Extracting maps of aligned to unaligned positions (i.e., gap maps)</a></li>
<li><a class="reference internal" href="#filtering-alignments-based-on-gaps">Filtering alignments based on gaps</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
              
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<div class="section" id="sequence-collections-and-alignments">
<h1>Sequence Collections and Alignments<a class="headerlink" href="#sequence-collections-and-alignments" title="Permalink to this headline">¶</a></h1>
<p>For loading collections of unaligned or aligned sequences see <a class="reference internal" href="loading_sequences.html#load-seqs"><span class="std std-ref">Loading an alignment from a file</span></a>.</p>
<div class="section" id="what-s-the-difference-between-alignment-and-arrayalignment">
<h2>What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?<a class="headerlink" href="#what-s-the-difference-between-alignment-and-arrayalignment" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> class can be annotated, meaning you can add annotations to an Alignment or it’s member sequences and you can slice the alignment via those objects. This capability is achieved, under the hood, by having the individual sequences represent gaps as a “span”, rather than explicitly as a “-” character in the sequence itself. This representation is also efficient for very long sequences.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> class cannot be annotated. The class represents its sequences as a <code class="docutils literal notranslate"><span class="pre">numpy.ndarray</span></code> instance. Thus, the gaps are included as a state in the array. This class is better at handling a lot of sequences and should typically be faster. This is the default class returned by the <code class="docutils literal notranslate"><span class="pre">load_aligned_seqs</span></code> and <code class="docutils literal notranslate"><span class="pre">make_aligned_seqs</span></code> functions. (See <a class="reference internal" href="loading_sequences.html#load-seqs"><span class="std std-ref">Loading an alignment from a file</span></a> for details.)</p>
<p>You can change alignment types using the <code class="docutils literal notranslate"><span class="pre">to_type()</span></code> method.</p>
</div>
<div class="section" id="basic-collection-objects">
<h2>Basic Collection objects<a class="headerlink" href="#basic-collection-objects" title="Permalink to this headline">¶</a></h2>
<div class="section" id="constructing-a-sequencecollection-or-alignment-object-from-strings">
<h3>Constructing a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from strings<a class="headerlink" href="#constructing-a-sequencecollection-or-alignment-object-from-strings" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span><span class="p">,</span> <span class="n">make_unaligned_seqs</span>

<span class="n">dna</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGACC&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATCGCC&quot;</span><span class="p">}</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dna</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">seqs</span><span class="p">))</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">make_unaligned_seqs</span><span class="p">(</span><span class="n">dna</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">seqs</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;cogent3.core.alignment.ArrayAlignment&#39;&gt;
&lt;class &#39;cogent3.core.alignment.SequenceCollection&#39;&gt;
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="constructing-a-arrayalignment-using-make-aligned-seqs">
<h3>Constructing a <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> using <code class="docutils literal notranslate"><span class="pre">make_aligned_seqs</span></code><a class="headerlink" href="#constructing-a-arrayalignment-using-make-aligned-seqs" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">dna</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGACC&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATCGCC&quot;</span><span class="p">}</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dna</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">seqs</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;cogent3.core.alignment.ArrayAlignment&#39;&gt;
&gt;seq1
ATGACC
&gt;seq2
ATCGCC

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="converting-a-sequencecollection-to-fasta-format">
<h3>Converting a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> to FASTA format<a class="headerlink" href="#converting-a-sequencecollection-to-fasta-format" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_unaligned_seqs</span>

<span class="n">seqs</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;NineBande
GCAAGGCGCCAACAGAGCAGATGGGCTGAAAGTAAGGAAACATGTAATGATAGGCAGACT
&gt;Mouse
GCAGTGAGCCAGCAGAGCAGATGGGCTGCAAGTAAAGGAACATGTAACGACAGGCAGGTT
&gt;Human
GCAAGGAGCCAACATAACAGATGGGCTGGAAGTAAGGAAACATGTAATGATAGGCGGACT
&gt;HowlerMon
GCAAGGAGCCAACATAACAGATGGGCTGAAAGTGAGGAAACATGTAATGATAGGCAGACT
&gt;DogFaced
GCAAGGAGCCAGCAGAACAGATGGGTTGAAACTAAGGAAACATGTAATGATAGGCAGACT

</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="adding-new-sequences-to-an-existing-collection-or-alignment">
<h2>Adding new sequences to an existing collection or alignment<a class="headerlink" href="#adding-new-sequences-to-an-existing-collection-or-alignment" title="Permalink to this headline">¶</a></h2>
<p>New sequences can be either appended or inserted using the <code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> method. More than one sequence can be added at the same time. Note that <code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> does not modify the existing collection/alignment, it creates a new one.</p>
<div class="section" id="appending-the-sequences">
<h3>Appending the sequences<a class="headerlink" href="#appending-the-sequences" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> without additional parameters will append the sequences to the end of the collection/alignment.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA------&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="n">new_seqs</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq0&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGT-AGG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq4&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGCC------&quot;</span><span class="p">)],</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span>
<span class="p">)</span>
<span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_seqs</span><span class="p">(</span><span class="n">new_seqs</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAA------
&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG

&gt;seq1
ATGAA------
&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG
&gt;seq0
ATG-AGT-AGG
&gt;seq4
ATGCC------

</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The order is not preserved if you use <code class="docutils literal notranslate"><span class="pre">to_fasta</span></code> method, which sorts sequences by name.</p>
</div>
</div>
<div class="section" id="inserting-the-sequences">
<h3>Inserting the sequences<a class="headerlink" href="#inserting-the-sequences" title="Permalink to this headline">¶</a></h3>
<p>Sequences can be inserted into an alignment at the specified position using either the <code class="docutils literal notranslate"><span class="pre">before_name</span></code> or <code class="docutils literal notranslate"><span class="pre">after_name</span></code> arguments.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_seqs</span><span class="p">(</span><span class="n">new_seqs</span><span class="p">,</span> <span class="n">before_name</span><span class="o">=</span><span class="s2">&quot;seq2&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
<span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_seqs</span><span class="p">(</span><span class="n">new_seqs</span><span class="p">,</span> <span class="n">after_name</span><span class="o">=</span><span class="s2">&quot;seq2&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAA------
&gt;seq0
ATG-AGT-AGG
&gt;seq4
ATGCC------
&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG

&gt;seq1
ATGAA------
&gt;seq2
ATG-AGTGATG
&gt;seq0
ATG-AGT-AGG
&gt;seq4
ATGCC------
&gt;seq3
AT--AG-GATG

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="inserting-sequence-s-based-on-their-alignment-to-a-reference-sequence">
<h3>Inserting sequence(s) based on their alignment to a reference sequence<a class="headerlink" href="#inserting-sequence-s-based-on-their-alignment-to-a-reference-sequence" title="Permalink to this headline">¶</a></h3>
<p>Already aligned sequences can be added to an existing <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object and aligned at the same time using the <code class="docutils literal notranslate"><span class="pre">add_from_ref_aln</span></code> method. The alignment is performed based on their alignment to a reference sequence (which must be present in both alignments). The method assumes the first sequence in <code class="docutils literal notranslate"><span class="pre">ref_aln.names[0]</span></code> is the reference.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA------&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ref_aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATAGGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq0&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGCG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq4&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGCTGGG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_from_ref_aln</span><span class="p">(</span><span class="n">ref_aln</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAA------
&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG
&gt;seq0
AT--G--AGCG
&gt;seq4
AT--GC-TGGG

</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">add_from_ref_aln</span></code> has the same arguments as <code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> so <code class="docutils literal notranslate"><span class="pre">before_name</span></code> and <code class="docutils literal notranslate"><span class="pre">after_name</span></code> can be used to insert the new sequences at the desired position.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method does not work with the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> class.</p>
</div>
<div class="section" id="removing-all-columns-with-gaps-in-a-named-sequence">
<h4>Removing all columns with gaps in a named sequence<a class="headerlink" href="#removing-all-columns-with-gaps-in-a-named-sequence" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA---TG-&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_degapped_relative_to</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAATG
&gt;seq2
ATG-AAT
&gt;seq3
AT--AAT

</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="the-elements-of-a-collection-or-alignment">
<h2>The elements of a collection or alignment<a class="headerlink" href="#the-elements-of-a-collection-or-alignment" title="Permalink to this headline">¶</a></h2>
<div class="section" id="accessing-individual-sequences-from-a-collection-or-alignment-by-name">
<h3>Accessing individual sequences from a collection or alignment by name<a class="headerlink" href="#accessing-individual-sequences-from-a-collection-or-alignment-by-name" title="Permalink to this headline">¶</a></h3>
<p>Using the <code class="docutils literal notranslate"><span class="pre">get_seq</span></code> method allows for extracting an unaligned sequence from a collection or alignment by name.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA------&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
    <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">name</span>
<span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">is_gapped</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>False
</pre></div>
</div>
</div>
</div>
<p>Alternatively, if you want to extract the aligned (i.e., gapped) sequence from an alignment, you can use <code class="docutils literal notranslate"><span class="pre">get_gapped_seq</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_gapped_seq</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">is_gapped</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>ATGAA------
</pre></div>
</div>
</div>
</div>
<p>To see the names of the sequences in a sequence collection, you can use either the <code class="docutils literal notranslate"><span class="pre">Names</span></code> attribute or <code class="docutils literal notranslate"><span class="pre">get_seq_names</span></code> method.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span><span class="o">.</span><span class="n">names</span>
<span class="n">aln</span><span class="o">.</span><span class="n">names</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;seq1&#39;, &#39;seq2&#39;, &#39;seq3&#39;]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="slice-the-sequences-from-an-alignment-like-a-list">
<h3>Slice the sequences from an alignment like a list<a class="headerlink" href="#slice-the-sequences-from-an-alignment-like-a-list" title="Permalink to this headline">¶</a></h3>
<p>The usual approach is to access a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object as a dictionary, obtaining the individual sequences using the titles as “keys” (above).  However, one can also iterate through the collection like a list.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_unaligned_seqs</span><span class="p">,</span> <span class="n">load_aligned_seqs</span>

<span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;data/long_testseqs.fasta&quot;</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">my_seq</span> <span class="o">=</span> <span class="n">seqs</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">my_seq</span><span class="p">[:</span><span class="mi">24</span><span class="p">]</span>
<span class="nb">str</span><span class="p">(</span><span class="n">my_seq</span><span class="p">[:</span><span class="mi">24</span><span class="p">])</span>
<span class="nb">type</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="mi">24</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="mi">24</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>TGTGGCACAAATACTCATGCCAGC
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="getting-a-subset-of-sequences-from-the-alignment">
<h3>Getting a subset of sequences from the alignment<a class="headerlink" href="#getting-a-subset-of-sequences-from-the-alignment" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">names</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">take_seqs</span><span class="p">([</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">])</span>
<span class="n">new</span><span class="o">.</span><span class="n">names</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;Human&#39;, &#39;HowlerMon&#39;]
</pre></div>
</div>
</div>
</div>
<p>Note, if you set <code class="docutils literal notranslate"><span class="pre">array_align=False</span></code>, then the subset contain references to the original sequences, not copies.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">,</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;Human&quot;</span><span class="p">)</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">take_seqs</span><span class="p">([</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">])</span>
<span class="nb">id</span><span class="p">(</span><span class="n">new</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;Human&quot;</span><span class="p">))</span> <span class="o">==</span> <span class="nb">id</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;Human&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="alignments">
<h2>Alignments<a class="headerlink" href="#alignments" title="Permalink to this headline">¶</a></h2>
<div class="section" id="creating-an-alignment-object-from-a-sequencecollection">
<h3>Creating an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code><a class="headerlink" href="#creating-an-alignment-object-from-a-sequencecollection" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_unaligned_seqs</span>
<span class="kn">from</span> <span class="nn">cogent3.core.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">fasta_1</span> <span class="o">=</span> <span class="n">seq</span>
<span class="n">fasta_2</span> <span class="o">=</span> <span class="n">aln</span>
<span class="k">assert</span> <span class="n">fasta_1</span> <span class="o">==</span> <span class="n">fasta_2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="convert-alignment-to-dna-rna-or-protein-moltypes">
<h3>Convert alignment to DNA, RNA or PROTEIN moltypes<a class="headerlink" href="#convert-alignment-to-dna-rna-or-protein-moltypes" title="Permalink to this headline">¶</a></h3>
<p>This is useful if you’ve loaded a sequence alignment without specifying the moltype and later need to convert it.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG---&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;CCTGGG&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="n">dna</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_dna</span><span class="p">()</span>
<span class="n">dna</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
td { border: none !important; text-align: left !important; }
.label { font-size: 12pt ; text-align: right !important; color: black !important; }
.T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<body>
<table>
<tr style="background-color:rgba(161, 195, 209, 0.5);"><td></td><td><b>0</b></td></tr>
<tr><td class="label">b</td><td><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span></td></tr>
<tr><td class="label">a</td><td><span class="A_dna">A</span><span class="C_dna">.</span><span class="G_dna">G</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span></td></tr>
</table>
<p><i>2 x 6 dna alignment</i></p>
</body></div></div>
</div>
<p>To RNA</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG---&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;CCUGGG&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="n">rna</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_rna</span><span class="p">()</span>
<span class="n">rna</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
td { border: none !important; text-align: left !important; }
.label { font-size: 12pt ; text-align: right !important; color: black !important; }
.U_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.C_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.A_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.G_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.terminal_ambig_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.ambig_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<body>
<table>
<tr style="background-color:rgba(161, 195, 209, 0.5);"><td></td><td><b>0</b></td></tr>
<tr><td class="label">b</td><td><span class="C_rna">C</span><span class="C_rna">C</span><span class="U_rna">U</span><span class="G_rna">G</span><span class="G_rna">G</span><span class="G_rna">G</span></td></tr>
<tr><td class="label">a</td><td><span class="A_rna">A</span><span class="C_rna">.</span><span class="G_rna">G</span><span class="ambig_rna">-</span><span class="ambig_rna">-</span><span class="ambig_rna">-</span></td></tr>
</table>
<p><i>2 x 6 rna alignment</i></p>
</body></div></div>
</div>
<p>To PROTEIN</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;TYV&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;TE-&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="n">prot</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_protein</span><span class="p">()</span>
<span class="n">prot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
td { border: none !important; text-align: left !important; }
.label { font-size: 12pt ; text-align: right !important; color: black !important; }
.A_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.C_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: orange; }
.D_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.E_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.F_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #ff6600; }
.G_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.H_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.I_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.K_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.L_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.M_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: orange; }
.N_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #993300; }
.P_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #cc0099; }
.Q_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #993300; }
.R_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.S_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009900; }
.T_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009900; }
.U_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.V_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.W_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #ff6600; }
.Y_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #ff6600; }
.terminal_ambig_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.ambig_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<body>
<table>
<tr style="background-color:rgba(161, 195, 209, 0.5);"><td></td><td><b>0</b></td></tr>
<tr><td class="label">x</td><td><span class="T_protein">T</span><span class="Y_protein">Y</span><span class="V_protein">V</span></td></tr>
<tr><td class="label">y</td><td><span class="T_protein">.</span><span class="E_protein">E</span><span class="ambig_protein">-</span></td></tr>
</table>
<p><i>2 x 3 protein alignment</i></p>
</body></div></div>
</div>
</div>
<div class="section" id="handling-gaps">
<h3>Handling gaps<a class="headerlink" href="#handling-gaps" title="Permalink to this headline">¶</a></h3>
<div class="section" id="remove-all-gaps-from-an-alignment-in-fasta-format">
<h4>Remove all gaps from an alignment in FASTA format<a class="headerlink" href="#remove-all-gaps-from-an-alignment-in-fasta-format" title="Permalink to this headline">¶</a></h4>
<p>This necessarily returns a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">)</span>
<span class="n">degapped</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">degap</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">degapped</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;cogent3.core.alignment.SequenceCollection&#39;&gt;
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="writing-sequences-to-file">
<h3>Writing sequences to file<a class="headerlink" href="#writing-sequences-to-file" title="Permalink to this headline">¶</a></h3>
<p>Both collection and alignment objects have a <code class="docutils literal notranslate"><span class="pre">write</span></code> method. The output format is inferred from the filename suffix,</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">dna</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGACC&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATCGCC&quot;</span><span class="p">}</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dna</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;sample.fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>or by the <code class="docutils literal notranslate"><span class="pre">format</span></code> argument.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.util.misc</span> <span class="kn">import</span> <span class="n">remove_files</span>

<span class="n">remove_files</span><span class="p">([</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;sample.fasta&quot;</span><span class="p">],</span> <span class="n">error_on_missing</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="converting-an-alignment-to-fasta-format">
<h3>Converting an alignment to FASTA format<a class="headerlink" href="#converting-an-alignment-to-fasta-format" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="kn">from</span> <span class="nn">cogent3.core.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">fasta_align</span> <span class="o">=</span> <span class="n">aln</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="converting-an-alignment-into-phylip-format">
<h3>Converting an alignment into Phylip format<a class="headerlink" href="#converting-an-alignment-into-phylip-format" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="kn">from</span> <span class="nn">cogent3.core.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">got</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_phylip</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">got</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>5  60
NineBande GCAAGGCGCCAACAGAGCAGATGGGCTGAAAGTAAGGAAACATGTAATGATAGGCAGACT
Mouse     GCAGTGAGCCAGCAGAGCAGATGGGCTGCAAGTAAAGGAACATGTAACGACAGGCAGGTT
Human     GCAAGGAGCCAACATAACAGATGGGCTGGAAGTAAGGAAACATGTAATGATAGGCGGACT
HowlerMon GCAAGGAGCCAACATAACAGATGGGCTGAAAGTGAGGAAACATGTAATGATAGGCAGACT
DogFaced  GCAAGGAGCCAGCAGAACAGATGGGTTGAAACTAAGGAAACATGTAATGATAGGCAGACT

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="converting-an-alignment-to-a-list-of-strings">
<h3>Converting an alignment to a list of strings<a class="headerlink" href="#converting-an-alignment-to-a-list-of-strings" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="kn">from</span> <span class="nn">cogent3.core.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">string_list</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
</div>
<div class="section" id="slicing-an-alignment">
<h2>Slicing an alignment<a class="headerlink" href="#slicing-an-alignment" title="Permalink to this headline">¶</a></h2>
<div class="section" id="by-rows-sequences">
<h3>By rows (sequences)<a class="headerlink" href="#by-rows-sequences" title="Permalink to this headline">¶</a></h3>
<p>An <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> can be sliced</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;data/long_testseqs.fasta&quot;</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="p">[:</span><span class="mi">24</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
TGTGGCACAAATACTCATGCCAGC
&gt;HowlerMon
TGTGGCACAAATACTCATGCCAGC
&gt;Mouse
TGTGGCACAGATGCTCATGCCAGC
&gt;NineBande
TGTGGCACAAATACTCATGCCAAC
&gt;DogFaced
TGTGGCACAAATACTCATGCCAAC

</pre></div>
</div>
</div>
</div>
<p>but a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> cannot be sliced</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_unaligned_seqs</span>

<span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;data/long_testseqs.fasta&quot;</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqs</span><span class="p">[:</span><span class="mi">24</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">TypeError</span><span class="g g-Whitespace">                                 </span>Traceback (most recent call last)
<span class="o">&lt;</span><span class="n">ipython</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">27</span><span class="o">-</span><span class="mi">703957846</span><span class="n">d13</span><span class="o">&gt;</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;data/long_testseqs.fasta&quot;</span>
<span class="g g-Whitespace">      </span><span class="mi">4</span> <span class="n">seqs</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span>
<span class="ne">----&gt; </span><span class="mi">5</span> <span class="nb">print</span><span class="p">(</span><span class="n">seqs</span><span class="p">[:</span><span class="mi">24</span><span class="p">])</span>

<span class="ne">TypeError</span>: &#39;SequenceCollection&#39; object is not subscriptable
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="getting-a-single-column-from-an-alignment">
<h3>Getting a single column from an alignment<a class="headerlink" href="#getting-a-single-column-from-an-alignment" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="n">column_four</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="getting-a-region-of-contiguous-columns">
<h3>Getting a region of contiguous columns<a class="headerlink" href="#getting-a-region-of-contiguous-columns" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">region</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">50</span><span class="p">:</span><span class="mi">70</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="iterating-over-alignment-positions">
<h3>Iterating over alignment positions<a class="headerlink" href="#iterating-over-alignment-positions" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">)</span>
<span class="n">col</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">113</span><span class="p">:</span><span class="mi">115</span><span class="p">]</span><span class="o">.</span><span class="n">iter_positions</span><span class="p">()</span>
<span class="nb">type</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
<span class="nb">list</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[[ByteSequence(A), ByteSequence(A), ByteSequence(A)],
 [ByteSequence(T), ByteSequence(-), ByteSequence(-)]]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="getting-codon-3rd-positions-from-alignment">
<h3>Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">Alignment</span></code><a class="headerlink" href="#getting-codon-3rd-positions-from-alignment" title="Permalink to this headline">¶</a></h3>
<p>We’ll do this by specifying the position indices of interest, creating a sequence <code class="docutils literal notranslate"><span class="pre">Feature</span></code> and using that to extract the positions.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGATGATG---&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGATGATGATG&quot;</span><span class="p">},</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aln</span><span class="p">))[</span><span class="mi">2</span><span class="p">::</span><span class="mi">3</span><span class="p">])</span>
<span class="n">indices</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aln</span><span class="p">))[</span><span class="mi">2</span><span class="p">::</span><span class="mi">3</span><span class="p">]]</span>
<span class="n">indices</span>
<span class="n">pos3</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;pos3&quot;</span><span class="p">,</span> <span class="s2">&quot;pos3&quot;</span><span class="p">,</span> <span class="n">indices</span><span class="p">)</span>
<span class="n">pos3</span> <span class="o">=</span> <span class="n">pos3</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pos3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
GGG-
&gt;seq2
GGGG

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="getting-codon-3rd-positions-from-arrayalignment">
<h3>Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code><a class="headerlink" href="#getting-codon-3rd-positions-from-arrayalignment" title="Permalink to this headline">¶</a></h3>
<p>We can use more conventional slice notation in this instance. Note, because Python counts from 0, the 3rd position starts at index 2.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGATGATG---&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGATGATGATG&quot;</span><span class="p">},</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">pos3</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">3</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pos3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
GGG-
&gt;seq2
GGGG

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="filtering-positions">
<span id="filter-positions"></span><h3>Filtering positions<a class="headerlink" href="#filtering-positions" title="Permalink to this headline">¶</a></h3>
<div class="section" id="trim-terminal-stop-codons">
<h4>Trim terminal stop codons<a class="headerlink" href="#trim-terminal-stop-codons" title="Permalink to this headline">¶</a></h4>
<p>For evolutionary analyses that use codon models we need to exclude terminating stop codons. For the case where the sequences are all of length divisible by 3.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGTAA---&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGACA---&quot;</span><span class="p">,</span> <span class="s2">&quot;seq3&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGCAATGA&quot;</span><span class="p">},</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">trim_stop_codons</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ACG---
&gt;seq2
ACGACA
&gt;seq3
ACGCAA

</pre></div>
</div>
</div>
</div>
<p>If the alignment contains sequences not divisible by 3, use the <code class="docutils literal notranslate"><span class="pre">allow_partial</span></code> argument.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGTAA---&quot;</span><span class="p">,</span>
        <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGAC----&quot;</span><span class="p">,</span>  <span class="c1"># terminal codon incomplete</span>
        <span class="s2">&quot;seq3&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGCAATGA&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">trim_stop_codons</span><span class="p">(</span><span class="n">allow_partial</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ACG---
&gt;seq2
ACGAC-
&gt;seq3
ACGCAA

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="eliminating-columns-with-non-nucleotide-characters">
<h4>Eliminating columns with non-nucleotide characters<a class="headerlink" href="#eliminating-columns-with-non-nucleotide-characters" title="Permalink to this headline">¶</a></h4>
<p>We sometimes want to eliminate ambiguous or gap data from our alignments. We show how to exclude alignment columns by the characters they contain. In the first instance we do this just for single nucleotide columns, then for trinucleotides (equivalent for handling codons). Both are done using the <code class="docutils literal notranslate"><span class="pre">no_degenerates</span></code> method.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTG---&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGNGATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>We apply to nucleotides,</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">nucs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">no_degenerates</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">nucs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAAGGG
&gt;seq2
ATGAAGGG
&gt;seq3
ATGAAGGG

</pre></div>
</div>
</div>
</div>
<p>Applying the same filter to trinucleotides (specified by setting <code class="docutils literal notranslate"><span class="pre">motif_length=3</span></code>).</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">trinucs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">no_degenerates</span><span class="p">(</span><span class="n">motif_length</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">trinucs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAAG
&gt;seq2
ATGAAG
&gt;seq3
ATGAAG

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="getting-all-variable-positions-from-an-alignment">
<h4>Getting all variable positions from an alignment<a class="headerlink" href="#getting-all-variable-positions-from-an-alignment" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">pos</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">variable_positions</span><span class="p">()</span>
<span class="n">just_variable_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">take_positions</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">just_variable_aln</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
AAGCAAAACT
&gt;HowlerMon
AAGCAAGACT
&gt;Mouse
GGGCCCAGCT
&gt;NineBande
AAATAAAACT
&gt;DogFaced
AAACAAAATA

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="getting-all-constant-positions-from-an-alignment">
<h4>Getting all constant positions from an alignment<a class="headerlink" href="#getting-all-constant-positions-from-an-alignment" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">pos</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">variable_positions</span><span class="p">()</span>
<span class="n">just_constant_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">take_positions</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">negate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">just_constant_aln</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
TGTGGCACAA
&gt;HowlerMon
TGTGGCACAA
&gt;Mouse
TGTGGCACAA
&gt;NineBande
TGTGGCACAA
&gt;DogFaced
TGTGGCACAA

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="getting-all-variable-codons-from-an-alignment">
<h4>Getting all variable codons from an alignment<a class="headerlink" href="#getting-all-variable-codons-from-an-alignment" title="Permalink to this headline">¶</a></h4>
<p>This is done using the <code class="docutils literal notranslate"><span class="pre">filtered</span></code> method using the <code class="docutils literal notranslate"><span class="pre">motif_length</span></code> argument. We demonstrate this first for the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">variable_codons</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">filtered</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">tuple</span><span class="p">,</span> <span class="n">x</span><span class="p">)))</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">,</span> <span class="n">motif_length</span><span class="o">=</span><span class="mi">3</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">just_variable_aln</span><span class="p">[:</span><span class="mi">9</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
AAGCAAAAC
&gt;HowlerMon
AAGCAAGAC
&gt;Mouse
GGGCCCAGC
&gt;NineBande
AAATAAAAC
&gt;DogFaced
AAACAAAAT

</pre></div>
</div>
</div>
</div>
<p>Then for the standard <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> by first converting the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_type</span><span class="p">(</span><span class="n">array_align</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">variable_codons</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">filtered</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">,</span> <span class="n">motif_length</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">just_variable_aln</span><span class="p">[:</span><span class="mi">9</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
AAGCAAAAC
&gt;HowlerMon
AAGCAAGAC
&gt;Mouse
GGGCCCAGC
&gt;NineBande
AAATAAAAC
&gt;DogFaced
AAACAAAAT

</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="filtering-sequences">
<h3>Filtering sequences<a class="headerlink" href="#filtering-sequences" title="Permalink to this headline">¶</a></h3>
<div class="section" id="extracting-sequences-by-sequence-identifier-into-a-new-alignment-object">
<h4>Extracting sequences by sequence identifier into a new alignment object<a class="headerlink" href="#extracting-sequences-by-sequence-identifier-into-a-new-alignment-object" title="Permalink to this headline">¶</a></h4>
<p>You can use <code class="docutils literal notranslate"><span class="pre">take_seqs</span></code> to extract some sequences by sequence identifier from an alignment to a new alignment object:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">take_seqs</span><span class="p">([</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;Mouse&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
td { border: none !important; text-align: left !important; }
.label { font-size: 12pt ; text-align: right !important; color: black !important; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.	_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.
_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
. _bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.!_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
."_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.#_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.$_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.%_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.&_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.'_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.(_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.)_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.*_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.+_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.,_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.-_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
./_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.0_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.1_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.2_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.3_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.4_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.5_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.6_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.7_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.8_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.9_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.:_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.;_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.<_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.=_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.>_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.?_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.@_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.A_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.B_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.C_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.D_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.E_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.F_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.G_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.H_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.I_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.J_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.K_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.L_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.M_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.N_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.O_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.P_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Q_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.R_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.S_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.T_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.U_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.V_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.W_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.X_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Y_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Z_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.[_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.\_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.]_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.^_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.__bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.`_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.a_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.b_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.d_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.e_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.f_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.g_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.h_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.i_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.j_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.k_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.l_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.m_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.n_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.o_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.p_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.q_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.r_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.s_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.t_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.u_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.v_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.w_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.x_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.y_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.z_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.{_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.|_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.}_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.~_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
. _bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¡_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¢_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.£_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¤_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¥_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¦_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.§_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¨_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.©_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ª_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.«_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¬_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.­_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.®_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¯_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.°_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.±_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.²_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.³_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.´_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.µ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¶_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.·_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¸_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¹_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.º_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.»_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¼_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.½_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¾_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¿_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.À_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Á_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Â_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ã_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ä_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Å_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Æ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ç_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.È_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.É_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ê_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ë_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ì_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Í_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Î_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ï_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ð_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ñ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ò_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ó_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ô_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Õ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ö_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.×_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ø_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ù_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ú_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Û_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ü_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ý_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Þ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ß_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.à_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.á_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.â_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ã_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ä_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.å_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.æ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ç_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.è_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.é_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ê_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ë_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ì_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.í_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.î_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ï_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ð_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ñ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ò_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ó_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ô_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.õ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ö_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.÷_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ø_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ù_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ú_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.û_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ü_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ý_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.þ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ÿ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.terminal_ambig_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ambig_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
</style>
<body>
<table>
<tr style="background-color:rgba(161, 195, 209, 0.5);"><td></td><td><b>0</b></td></tr>
<tr><td class="label">Mouse</td><td><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="T_bytes">T</span></td></tr>
<tr><td class="label">Human</td><td><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">A</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">A</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">A</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">A</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">A</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">C</span><span class="T_bytes">.</span></td></tr>
</table>
<p><i>2 x 2532 (truncated to 2 x 60) bytes alignment</i></p>
</body></div></div>
</div>
<p>Alternatively, you can extract only the sequences which are not specified by passing <code class="docutils literal notranslate"><span class="pre">negate=True</span></code>:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span><span class="o">.</span><span class="n">take_seqs</span><span class="p">([</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;Mouse&quot;</span><span class="p">],</span> <span class="n">negate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
td { border: none !important; text-align: left !important; }
.label { font-size: 12pt ; text-align: right !important; color: black !important; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.	_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.
_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
. _bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.!_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
."_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.#_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.$_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.%_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.&_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.'_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.(_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.)_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.*_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.+_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.,_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.-_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
./_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.0_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.1_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.2_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.3_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.4_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.5_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.6_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.7_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.8_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.9_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.:_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.;_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.<_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.=_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.>_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.?_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.@_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.A_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.B_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.C_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.D_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.E_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.F_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.G_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.H_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.I_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.J_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.K_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.L_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.M_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.N_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.O_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.P_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Q_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.R_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.S_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.T_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.U_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.V_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.W_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.X_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Y_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Z_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.[_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.\_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.]_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.^_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.__bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.`_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.a_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.b_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.d_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.e_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.f_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.g_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.h_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.i_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.j_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.k_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.l_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.m_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.n_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.o_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.p_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.q_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.r_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.s_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.t_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.u_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.v_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.w_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.x_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.y_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.z_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.{_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.|_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.}_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.~_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
. _bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¡_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¢_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.£_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¤_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¥_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¦_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.§_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¨_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.©_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ª_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.«_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¬_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.­_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.®_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¯_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.°_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.±_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.²_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.³_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.´_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.µ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¶_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.·_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¸_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¹_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.º_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.»_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¼_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.½_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¾_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.¿_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.À_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Á_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Â_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ã_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ä_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Å_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Æ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ç_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.È_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.É_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ê_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ë_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ì_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Í_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Î_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ï_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ð_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ñ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ò_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ó_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ô_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Õ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ö_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.×_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ø_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ù_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ú_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Û_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ü_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Ý_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.Þ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ß_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.à_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.á_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.â_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ã_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ä_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.å_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.æ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ç_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.è_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.é_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ê_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ë_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ì_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.í_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.î_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ï_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ð_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ñ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ò_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ó_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ô_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.õ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ö_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.÷_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ø_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ù_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ú_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.û_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ü_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ý_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.þ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ÿ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.terminal_ambig_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.ambig_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
</style>
<body>
<table>
<tr style="background-color:rgba(161, 195, 209, 0.5);"><td></td><td><b>0</b></td></tr>
<tr><td class="label">NineBande</td><td><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span></td></tr>
<tr><td class="label">HowlerMon</td><td><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">G</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">C</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">G</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span></td></tr>
<tr><td class="label">DogFaced</td><td><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">C</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span></td></tr>
</table>
<p><i>3 x 2532 (truncated to 3 x 60) bytes alignment</i></p>
</body></div></div>
</div>
</div>
<div class="section" id="extracting-sequences-using-an-arbitrary-function-into-a-new-alignment-object">
<h4>Extracting sequences using an arbitrary function into a new alignment object<a class="headerlink" href="#extracting-sequences-using-an-arbitrary-function-into-a-new-alignment-object" title="Permalink to this headline">¶</a></h4>
<p>You can use <code class="docutils literal notranslate"><span class="pre">take_seqs_if</span></code> to extract sequences into a new alignment object based on whether an arbitrary function applied to the sequence evaluates to True. For example, to extract sequences which don’t contain any N bases you could do the following:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTG---&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGNGATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="k">def</span> <span class="nf">no_N_chars</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
    <span class="k">return</span> <span class="s2">&quot;N&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">s</span>

<span class="n">aln</span><span class="o">.</span><span class="n">take_seqs_if</span><span class="p">(</span><span class="n">no_N_chars</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
td { border: none !important; text-align: left !important; }
.label { font-size: 12pt ; text-align: right !important; color: black !important; }
.T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<body>
<table>
<tr style="background-color:rgba(161, 195, 209, 0.5);"><td></td><td><b>0</b></td></tr>
<tr><td class="label">seq2</td><td><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span></td></tr>
<tr><td class="label">seq1</td><td><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span></td></tr>
</table>
<p><i>2 x 12 dna alignment</i></p>
</body></div></div>
</div>
<p>You can additionally get the sequences where the provided function evaluates to False:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span><span class="o">.</span><span class="n">take_seqs_if</span><span class="p">(</span><span class="n">no_N_chars</span><span class="p">,</span> <span class="n">negate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
td { border: none !important; text-align: left !important; }
.label { font-size: 12pt ; text-align: right !important; color: black !important; }
.T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<body>
<table>
<tr style="background-color:rgba(161, 195, 209, 0.5);"><td></td><td><b>0</b></td></tr>
<tr><td class="label">seq3</td><td><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="ambig_dna">N</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span></td></tr>
</table>
<p><i>1 x 12 dna alignment</i></p>
</body></div></div>
</div>
</div>
</div>
</div>
<div class="section" id="computing-alignment-statistics">
<h2>Computing alignment statistics<a class="headerlink" href="#computing-alignment-statistics" title="Permalink to this headline">¶</a></h2>
<div class="section" id="getting-motif-counts">
<h3>Getting motif counts<a class="headerlink" href="#getting-motif-counts" title="Permalink to this headline">¶</a></h3>
<p>We state the motif length we want and whether to allow gap or ambiguous characters. The latter only has meaning for IPUAC character sets (the DNA, RNA or PROTEIN moltypes). We illustrate this for the DNA moltype with motif lengths of 1 and 3.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTG---&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGNGATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">counts</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">counts</span><span class="p">(</span><span class="n">motif_length</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">counts</span><span class="p">(</span><span class="n">include_ambiguity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>==================
 A    C     G    T
------------------
11    0    14    7
------------------
============================================================================================================================================================================================================================================================================================================================================================================================================================================================
AAA    AAC    AAG    AAT    ACA    ACC    ACG    ACT    AGA    AGC    AGG    AGT    ATA    ATC    ATG    ATT    CAA    CAC    CAG    CAT    CCA    CCC    CCG    CCT    CGA    CGC    CGG    CGT    CTA    CTC    CTG    CTT    GAA    GAC    GAG    GAT    GCA    GCC    GCG    GCT    GGA    GGC    GGG    GGT    GTA    GTC    GTG    GTT    TAA    TAC    TAG    TAT    TCA    TCC    TCG    TCT    TGA    TGC    TGG    TGT    TTA    TTC    TTG    TTT
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  0      0      3      0      0      0      0      0      0      0      0      0      0      0      5      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      2      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
=======================
 A    C     G    N    T
-----------------------
11    0    14    1    7
-----------------------
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only the observed motifs are returned, rather than all defined by the alphabet.</p>
</div>
</div>
<div class="section" id="computing-motif-probabilities-from-an-alignment">
<h3>Computing motif probabilities from an alignment<a class="headerlink" href="#computing-motif-probabilities-from-an-alignment" title="Permalink to this headline">¶</a></h3>
<p>The method <code class="docutils literal notranslate"><span class="pre">get_motif_probs</span></code> of <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects returns the probabilities for all motifs of a given length. For individual nucleotides:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">motif_probs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>{&#39;T&#39;: 0.25520602569782896, &#39;C&#39;: 0.25808595480726626, &#39;A&#39;: 0.24390784226849802, &#39;G&#39;: 0.24280017722640673}
</pre></div>
</div>
</div>
</div>
<p>For dinucleotides or longer, we need to pass in an <code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> with the appropriate word length. Here is an example with trinucleotides:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span><span class="p">,</span> <span class="n">DNA</span>

<span class="n">trinuc_alphabet</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">get_word_alphabet</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">(</span><span class="n">alphabet</span><span class="o">=</span><span class="n">trinuc_alphabet</span><span class="p">)</span>
<span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">motif_probs</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">motif_probs</span><span class="p">[</span><span class="n">x</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">  </span><span class="si">%.3f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">motif_probs</span><span class="p">[</span><span class="n">m</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>CAG  0.037
CCT  0.034
CGC  0.030
TCT  0.029
GCC  0.028
AAC  0.028
AAA  0.027
AAG  0.027
GGG  0.026
CCC  0.025
TAA  0.025
CTT  0.024
TTC  0.024
GAG  0.024
TGT  0.023
CTG  0.023
CTC  0.021
GGA  0.021
TTT  0.020
GGT  0.019
CAA  0.019
AGT  0.019
TCC  0.019
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>
GAA  0.019
ATT  0.018
TAG  0.017
AAT  0.017
TTG  0.017
TCA  0.015
TGG  0.015
TGC  0.015
GTT  0.014
GGC  0.014
CAT  0.013
GTA  0.013
TGA  0.013
ATA  0.012
AGA  0.012
CAC  0.011
TTA  0.011
CCA  0.011
GCT  0.011
GAT  0.011
GCA  0.011
CCG  0.011
TAC  0.010
GAC  0.009
AGC  0.009
ATC  0.008
GTC  0.008
CGG  0.008
AGG  0.008
ACT  0.007
ACG  0.007
CGT  0.007
GCG  0.007
GTG  0.006
TAT  0.005
ATG  0.005
TCG  0.005
ACC  0.005
CGA  0.005
CTA  0.003
ACA  0.002
</pre></div>
</div>
</div>
</div>
<p>The same holds for other arbitrary alphabets, as long as they match the alignment <code class="docutils literal notranslate"><span class="pre">MolType</span></code>.</p>
<p>Some calculations in <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> require all non-zero values in the motif probabilities, in which case we use a pseudo-count. We illustrate that here with a simple example where T is missing. Without the pseudo-count, the frequency of T is 0.0, with the pseudo-count defined as 1e-6 then the frequency of T will be slightly less than 1e-6.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">[(</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;AACAAC&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;AAGAAG&quot;</span><span class="p">)],</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">()</span>
<span class="k">assert</span> <span class="n">motif_probs</span><span class="p">[</span><span class="s2">&quot;T&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mf">0.0</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">(</span><span class="n">pseudocount</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">)</span>
<span class="k">assert</span> <span class="mi">0</span> <span class="o">&lt;</span> <span class="n">motif_probs</span><span class="p">[</span><span class="s2">&quot;T&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">1e-6</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>It is important to notice that motif probabilities are computed by treating sequences as non-overlapping tuples. Below is a very simple pair of identical sequences where there are clearly 2 ‘AA’ dinucleotides per sequence but only the first one is ‘in-frame’ (frame width = 2).</p>
<p>We then create a dinucleotide <code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> object and use this to get dinucleotide probabilities. These frequencies are determined by breaking each aligned sequence up into non-overlapping dinucleotides and then doing a count. The expected value for the ‘AA’ dinucleotide in this case will be 2/8 = 0.25.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;AACGTAAG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;AACGTAAG&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">dinuc_alphabet</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">get_word_alphabet</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">(</span><span class="n">alphabet</span><span class="o">=</span><span class="n">dinuc_alphabet</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">motif_probs</span><span class="p">[</span><span class="s2">&quot;AA&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mf">0.25</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>What about counting the total incidence of dinucleotides including those not in-frame?  A naive application of the Python string object’s count method will not work as desired either because it “returns the number of non-overlapping occurrences”.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;my_seq&quot;</span><span class="p">,</span> <span class="s2">&quot;AAAGTAAG&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">my_seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;my_seq&quot;</span><span class="p">)</span>
<span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="s2">&quot;AAA&quot;</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="s2">&quot;AAAA&quot;</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>2
</pre></div>
</div>
</div>
</div>
<p>To count all occurrences of a given dinucleotide in a DNA sequence, one could use a standard Python approach such as list comprehension:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">make_seq</span><span class="p">(</span><span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span> <span class="n">seq</span><span class="o">=</span><span class="s2">&quot;AAAGTAAG&quot;</span><span class="p">)</span>
<span class="n">seq</span>
<span class="n">di_nucs</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span> <span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">2</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]</span>
<span class="nb">sum</span><span class="p">([</span><span class="n">nn</span> <span class="o">==</span> <span class="s2">&quot;AA&quot;</span> <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="n">di_nucs</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>3
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="working-with-alignment-gaps">
<h3>Working with alignment gaps<a class="headerlink" href="#working-with-alignment-gaps" title="Permalink to this headline">¶</a></h3>
<div class="section" id="filtering-extracted-columns-for-the-gap-character">
<h4>Filtering extracted columns for the gap character<a class="headerlink" href="#filtering-extracted-columns-for-the-gap-character" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">)</span>
<span class="n">col</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">113</span><span class="p">:</span><span class="mi">115</span><span class="p">]</span><span class="o">.</span><span class="n">iter_positions</span><span class="p">()</span>
<span class="n">c1</span><span class="p">,</span> <span class="n">c2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
<span class="n">c1</span><span class="p">,</span> <span class="n">c2</span>
<span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">c1</span><span class="p">))</span>
<span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">c2</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[ByteSequence(-), ByteSequence(-)]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="calculating-the-gap-fraction">
<h4>Calculating the gap fraction<a class="headerlink" href="#calculating-the-gap-fraction" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">aln</span><span class="p">[</span><span class="mi">113</span><span class="p">:</span><span class="mi">150</span><span class="p">]</span><span class="o">.</span><span class="n">iter_positions</span><span class="p">():</span>
    <span class="n">ungapped</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">column</span><span class="p">))</span>
    <span class="n">gap_fraction</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ungapped</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">gap_fraction</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>0.0
0.6666666666666666
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="extracting-maps-of-aligned-to-unaligned-positions-i-e-gap-maps">
<h4>Extracting maps of aligned to unaligned positions (i.e., gap maps)<a class="headerlink" href="#extracting-maps-of-aligned-to-unaligned-positions-i-e-gap-maps" title="Permalink to this headline">¶</a></h4>
<p>It’s often important to know how an alignment position relates to a position in one or more of the sequences in the alignment. The <code class="docutils literal notranslate"><span class="pre">gap_maps</span></code> method of the individual sequences is useful for this. To get a map of sequence to alignment positions for a specific sequence in your alignment, do the following:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGG-TG--&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAG--GATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">seq_to_aln_map</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_gapped_seq</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">gap_maps</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>It’s now possible to look up positions in the <code class="docutils literal notranslate"><span class="pre">seq1</span></code>, and find out what they map to in the alignment:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seq_to_aln_map</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">seq_to_aln_map</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>9
</pre></div>
</div>
</div>
</div>
<p>This tells us that in position 3 in <code class="docutils literal notranslate"><span class="pre">seq1</span></code> corresponds to position 3 in <code class="docutils literal notranslate"><span class="pre">aln</span></code>, and that position 8 in <code class="docutils literal notranslate"><span class="pre">seq1</span></code> corresponds to position 9 in <code class="docutils literal notranslate"><span class="pre">aln</span></code>.</p>
<p>Notice that we grabbed the first result from the call to <code class="docutils literal notranslate"><span class="pre">gap_maps</span></code>. This is the sequence position to alignment position map. The second value returned is the alignment position to sequence position map, so if you want to find out what sequence positions the alignment positions correspond to (opposed to what alignment positions the sequence positions correspond to) for a given sequence, you would take the following steps:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln_to_seq_map</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_gapped_seq</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">gap_maps</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">aln_to_seq_map</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">aln_to_seq_map</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>7
</pre></div>
</div>
</div>
</div>
<p>If an alignment position is a gap, and therefore has no corresponding sequence position, you’ll get a <code class="docutils literal notranslate"><span class="pre">KeyError</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seq_pos</span> <span class="o">=</span> <span class="n">aln_to_seq_map</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">KeyError</span><span class="g g-Whitespace">                                  </span>Traceback (most recent call last)
<span class="o">&lt;</span><span class="n">ipython</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">58</span><span class="o">-</span><span class="mi">76706</span><span class="n">f04c62d</span><span class="o">&gt;</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">seq_pos</span> <span class="o">=</span> <span class="n">aln_to_seq_map</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>

<span class="ne">KeyError</span>: 7
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The first position in alignments and sequences is always numbered position 0.</p>
</div>
</div>
<div class="section" id="filtering-alignments-based-on-gaps">
<h4>Filtering alignments based on gaps<a class="headerlink" href="#filtering-alignments-based-on-gaps" title="Permalink to this headline">¶</a></h4>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>An alternate, computationally faster, approach to removing gaps is to use the <code class="docutils literal notranslate"><span class="pre">filtered</span></code> method as discussed in <a class="reference internal" href="#filter-positions"><span class="std std-ref">Filtering positions</span></a>.</p>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">omit_gap_runs</span></code> method can be applied to remove long stretches of gaps in an alignment. In the following example, we remove sequences that have more than two adjacent gaps anywhere in the aligned sequence.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA---TG-&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">omit_gap_runs</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG

</pre></div>
</div>
</div>
</div>
<p>If instead, we just wanted to remove positions from the alignment which are gaps in more than a certain percentage of the sequences, we could use the <code class="docutils literal notranslate"><span class="pre">omit_gap_pos</span></code> function. For example:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA---TG-&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">omit_gap_pos</span><span class="p">(</span><span class="mf">0.40</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGA--TG-
&gt;seq2
ATGAGGATG
&gt;seq3
AT-AGGATG

</pre></div>
</div>
</div>
</div>
<p>If you wanted to remove sequences which contain more than a certain percent gap characters, you could use the <code class="docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code> method. This is commonly applied to filter partial sequences from an alignment.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA------&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">filtered_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">omit_gap_seqs</span><span class="p">(</span><span class="mf">0.50</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">filtered_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG

</pre></div>
</div>
</div>
</div>
<p>Note that following this call to <code class="docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code>, the 4th column of <code class="docutils literal notranslate"><span class="pre">filtered_aln</span></code> is 100% gaps. This is generally not desirable, so a call to <code class="docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code> is frequently followed with a call to <code class="docutils literal notranslate"><span class="pre">omit_gap_pos</span></code> with no parameters – this defaults to removing positions which are all gaps:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">filtered_aln</span><span class="o">.</span><span class="n">omit_gap_pos</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq2
ATGAGTGATG
&gt;seq3
AT-AG-GATG
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020, cogent3 Team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.1.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>