<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Evolutionary Analysis Using Likelihood &#8212; cogent3 2020.4 documentation</title>
    <link rel="stylesheet" href="../../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../_static/require.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../../_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">
          cogent3</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../install.html">Install</a></li>
                <li><a href="../index.html">Docs</a></li>
                <li><a href="../draw/index.html">Gallery</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Project History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pycogent.html">How does <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> relate to PyCogent?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Evolutionary Analysis Using Likelihood</a><ul>
<li><a class="reference internal" href="#specifying-substitution-models">Specifying substitution models</a><ul>
<li><a class="reference internal" href="#the-available-pre-defined-substitution-models">The available pre-defined substitution models</a></li>
<li><a class="reference internal" href="#getting-a-substitution-model-with-get-model">Getting a substitution model with <code class="docutils literal notranslate"><span class="pre">get_model()</span></code></a></li>
<li><a class="reference internal" href="#rate-heterogeneity-models">Rate heterogeneity models</a><ul>
<li><a class="reference internal" href="#for-nucleotide">For nucleotide</a></li>
<li><a class="reference internal" href="#for-codon">For codon</a></li>
<li><a class="reference internal" href="#for-protein">For protein</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#making-a-likelihood-function">Making a likelihood function</a><ul>
<li><a class="reference internal" href="#providing-an-alignment-to-a-likelihood-function">Providing an alignment to a likelihood function</a></li>
<li><a class="reference internal" href="#scoping-parameters-on-trees-time-heterogeneous-models">Scoping parameters on trees – time heterogeneous models</a></li>
<li><a class="reference internal" href="#specifying-a-parameter-as-constant">Specifying a parameter as constant</a></li>
<li><a class="reference internal" href="#providing-a-starting-value-for-a-parameter">Providing a starting value for a parameter</a></li>
<li><a class="reference internal" href="#setting-parameter-bounds-for-optimisation">Setting parameter bounds for optimisation</a></li>
<li><a class="reference internal" href="#setting-an-upper-bound-for-branch-length">Setting an upper bound for branch length</a></li>
<li><a class="reference internal" href="#specifying-rate-heterogeneity-functions">Specifying rate heterogeneity functions</a></li>
<li><a class="reference internal" href="#specifying-phylo-hmms">Specifying Phylo-HMMs</a></li>
<li><a class="reference internal" href="#fitting-likelihood-functions-choice-of-optimisers">Fitting likelihood functions - Choice of optimisers</a></li>
<li><a class="reference internal" href="#how-to-check-your-optimisation-was-successful">How to check your optimisation was successful</a></li>
<li><a class="reference internal" href="#overview-of-the-fitted-likelihood-function">Overview of the fitted likelihood function</a><ul>
<li><a class="reference internal" href="#log-likelihood-and-number-of-free-parameters">Log likelihood and number of free parameters</a></li>
<li><a class="reference internal" href="#aikake-information-criterion">Aikake Information Criterion</a></li>
<li><a class="reference internal" href="#bayesian-information-criterion">Bayesian Information Criterion</a></li>
<li><a class="reference internal" href="#getting-maximum-likelihood-estimates">Getting maximum likelihood estimates</a><ul>
<li><a class="reference internal" href="#one-at-a-time">One at a time</a></li>
<li><a class="reference internal" href="#just-the-motif-probabilities">Just the motif probabilities</a></li>
<li><a class="reference internal" href="#as-tables">As tables</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#testing-hypotheses-using-likelihood-ratio-tests">Testing Hypotheses - Using Likelihood Ratio Tests</a><ul>
<li><a class="reference internal" href="#testing-hypotheses-by-parametric-bootstrapping">Testing Hypotheses - By parametric bootstrapping</a></li>
</ul>
</li>
<li><a class="reference internal" href="#determining-confidence-intervals-on-mles">Determining confidence intervals on MLEs</a></li>
<li><a class="reference internal" href="#saving-results">Saving results</a></li>
<li><a class="reference internal" href="#reconstructing-ancestral-sequences">Reconstructing ancestral sequences</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
              
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="section" id="evolutionary-analysis-using-likelihood">
<h1>Evolutionary Analysis Using Likelihood<a class="headerlink" href="#evolutionary-analysis-using-likelihood" title="Permalink to this headline">¶</a></h1>
<div class="section" id="specifying-substitution-models">
<h2>Specifying substitution models<a class="headerlink" href="#specifying-substitution-models" title="Permalink to this headline">¶</a></h2>
<div class="section" id="the-available-pre-defined-substitution-models">
<h3>The available pre-defined substitution models<a class="headerlink" href="#the-available-pre-defined-substitution-models" title="Permalink to this headline">¶</a></h3>
<p>In cases where code takes a substitution model as an argument, you can use the value under “Abbreviation” as a string.</p>
</div>
<div class="section" id="getting-a-substitution-model-with-get-model">
<h3>Getting a substitution model with <code class="docutils literal notranslate"><span class="pre">get_model()</span></code><a class="headerlink" href="#getting-a-substitution-model-with-get-model" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="rate-heterogeneity-models">
<h3>Rate heterogeneity models<a class="headerlink" href="#rate-heterogeneity-models" title="Permalink to this headline">¶</a></h3>
<p>We illustrate this for the gamma distributed case using examples of the canned models displayed above. Creating rate heterogeneity variants of the canned models can be done by using optional arguments that get passed to the substitution model class.</p>
<div class="section" id="for-nucleotide">
<h4>For nucleotide<a class="headerlink" href="#for-nucleotide" title="Permalink to this headline">¶</a></h4>
<p>We specify a general time reversible nucleotide model with gamma distributed rate heterogeneity.</p>
</div>
<div class="section" id="for-codon">
<h4>For codon<a class="headerlink" href="#for-codon" title="Permalink to this headline">¶</a></h4>
<p>We specify a conditional nucleotide frequency codon model with nucleotide general time reversible parameters and a parameter for the ratio of nonsynonymous to synonymous substitutions (omega) with gamma distributed rate heterogeneity.</p>
</div>
<div class="section" id="for-protein">
<h4>For protein<a class="headerlink" href="#for-protein" title="Permalink to this headline">¶</a></h4>
<p>We specify a Jones, Taylor and Thornton 1992 empirical protein substitution model with gamma distributed rate heterogeneity.</p>
</div>
</div>
</div>
<div class="section" id="making-a-likelihood-function">
<h2>Making a likelihood function<a class="headerlink" href="#making-a-likelihood-function" title="Permalink to this headline">¶</a></h2>
<p>You start by specifying a substitution model and use that to construct a likelihood function for a specific tree.</p>
<div class="section" id="providing-an-alignment-to-a-likelihood-function">
<h3>Providing an alignment to a likelihood function<a class="headerlink" href="#providing-an-alignment-to-a-likelihood-function" title="Permalink to this headline">¶</a></h3>
<p>You need to load an alignment and then provide it a likelihood function. I construct very simple trees and alignments for this example.</p>
</div>
<div class="section" id="scoping-parameters-on-trees-time-heterogeneous-models">
<h3>Scoping parameters on trees – time heterogeneous models<a class="headerlink" href="#scoping-parameters-on-trees-time-heterogeneous-models" title="Permalink to this headline">¶</a></h3>
<p>For many evolutionary analyses, it’s desirable to allow different branches on a tree to have different values of a parameter. We show this for a simple codon model case here where we want the great apes (the clade that includes human and orangutan) to have a different value of the ratio of nonsynonymous to synonymous substitutions. This parameter is identified in the precanned <code class="docutils literal notranslate"><span class="pre">CNFGTR</span></code> model as <code class="docutils literal notranslate"><span class="pre">omega</span></code>.</p>
<p>We’ve set an <em>initial</em> value for this clade so that the edges affected by this rule are evident below.</p>
<p>A more extensive description of capabilities is in <a class="reference internal" href="../examples/scope_model_params_on_trees.html#scope-params-on-trees"><span class="std std-ref">Allowing substitution model parameters to differ between branches</span></a>.</p>
</div>
<div class="section" id="specifying-a-parameter-as-constant">
<h3>Specifying a parameter as constant<a class="headerlink" href="#specifying-a-parameter-as-constant" title="Permalink to this headline">¶</a></h3>
<p>This means the parameter will not be modified during likelihood maximisation. We show this here by making the <code class="docutils literal notranslate"><span class="pre">omega</span></code> parameter constant at the value 1 – essentially the condition of selective neutrality.</p>
</div>
<div class="section" id="providing-a-starting-value-for-a-parameter">
<h3>Providing a starting value for a parameter<a class="headerlink" href="#providing-a-starting-value-for-a-parameter" title="Permalink to this headline">¶</a></h3>
<p>This can be useful to improve performance, the closer you are to the maximum likelihood estimator the quicker optimisation will be.</p>
</div>
<div class="section" id="setting-parameter-bounds-for-optimisation">
<h3>Setting parameter bounds for optimisation<a class="headerlink" href="#setting-parameter-bounds-for-optimisation" title="Permalink to this headline">¶</a></h3>
<p>This can be useful for stopping optimisers from getting stuck in a bad part of parameter space. The following is for <code class="docutils literal notranslate"><span class="pre">omega</span></code> in a codon model. I’m also providing an initial guess for the parameter (<code class="docutils literal notranslate"><span class="pre">init=0.1</span></code>) as well as a lower bound. An initial guess that is close to the maximum likelihood estimate will speed up optimisation.</p>
</div>
<div class="section" id="setting-an-upper-bound-for-branch-length">
<h3>Setting an upper bound for branch length<a class="headerlink" href="#setting-an-upper-bound-for-branch-length" title="Permalink to this headline">¶</a></h3>
<p>If the branch length estimates seem too large, setting just an upper bound can be sensible. This will apply to all edges on the tree.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If, after optimising, the branch lengths equal to the upper value you set then the function has not been fully maximised and you should consider adjusting the boundary again.</p>
</div>
</div>
<div class="section" id="specifying-rate-heterogeneity-functions">
<h3>Specifying rate heterogeneity functions<a class="headerlink" href="#specifying-rate-heterogeneity-functions" title="Permalink to this headline">¶</a></h3>
<p>We extend the simple gamma distributed rate heterogeneity case for nucleotides from above to construction of the actual likelihood function. We do this for 4 bins and constraint the bin probabilities to be equal.</p>
<p>For more detailed discussion of defining and using these models see <a class="reference internal" href="../examples/rate_heterogeneity.html#rate-heterogeneity"><span class="std std-ref">Analysis of rate heterogeneity</span></a>.</p>
</div>
<div class="section" id="specifying-phylo-hmms">
<h3>Specifying Phylo-HMMs<a class="headerlink" href="#specifying-phylo-hmms" title="Permalink to this headline">¶</a></h3>
<p>For more detailed discussion of defining and using these models see <a class="reference internal" href="../examples/hmm_par_heterogeneity.html#rate-heterogeneity-hmm"><span class="std std-ref">Evaluate process heterogeneity using a Hidden Markov Model</span></a>.</p>
</div>
<div class="section" id="fitting-likelihood-functions-choice-of-optimisers">
<h3>Fitting likelihood functions - Choice of optimisers<a class="headerlink" href="#fitting-likelihood-functions-choice-of-optimisers" title="Permalink to this headline">¶</a></h3>
<p>There are 2 types of optimiser: simulated annealing, a <em>global</em> optimiser; and Powell, a <em>local</em> optimiser. The simulated annealing method is slow compared to Powell and in general Powell is an adequate choice. I setup a simple nucleotide model to illustrate these.</p>
<p>The default is to use Powell. For Powell, it’s recommended to set the <code class="docutils literal notranslate"><span class="pre">max_restarts</span></code> argument since this provides a mechanism for Powell to attempt restarting the optimisation from a slightly different spot which can help in overcoming local maxima.</p>
<p>We might want to do crude simulated annealing following by more rigorous Powell. To do this we first need to use the global optimiser, setting <code class="docutils literal notranslate"><span class="pre">local=False</span></code> setting a large value for <code class="docutils literal notranslate"><span class="pre">global_tolerance</span></code>.</p>
<p>Followed by a standard call to <code class="docutils literal notranslate"><span class="pre">optimise()</span></code>.</p>
</div>
<div class="section" id="how-to-check-your-optimisation-was-successful">
<h3>How to check your optimisation was successful<a class="headerlink" href="#how-to-check-your-optimisation-was-successful" title="Permalink to this headline">¶</a></h3>
<p>There is no guarantee that an optimised function has achieved a global maximum. We can, however, be sure that a maximum was achieved by validating that the optimiser stopped because the specified tolerance condition was met, rather than exceeding the maximum number of evaluations. The latter number is set to ensure optimisation doesn’t proceed endlessly. If the optimiser exited because this limit was exceeded you can be sure that the function <strong>has not</strong> been successfully optimised.</p>
<p>We can monitor this situation using the <code class="docutils literal notranslate"><span class="pre">limit_action</span></code> argument to <code class="docutils literal notranslate"><span class="pre">optimise</span></code>. Providing the value <code class="docutils literal notranslate"><span class="pre">raise</span></code> causes an exception to be raised if this condition occurs, as shown below. Providing <code class="docutils literal notranslate"><span class="pre">warn</span></code> (default) instead will cause a warning message to be printed to screen but execution will continue. The value <code class="docutils literal notranslate"><span class="pre">ignore</span></code> hides any such message.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We recommend using <code class="docutils literal notranslate"><span class="pre">limit_action='raise'</span></code> and catching the <code class="docutils literal notranslate"><span class="pre">ArithmeticError</span></code> error explicitly (as demonstrated above). You really shouldn’t be using results from such an optimisation run.</p>
</div>
</div>
<div class="section" id="overview-of-the-fitted-likelihood-function">
<h3>Overview of the fitted likelihood function<a class="headerlink" href="#overview-of-the-fitted-likelihood-function" title="Permalink to this headline">¶</a></h3>
<p>In Jupyter, the likelihood function object presents a representation of the main object features.</p>
<div class="section" id="log-likelihood-and-number-of-free-parameters">
<h4>Log likelihood and number of free parameters<a class="headerlink" href="#log-likelihood-and-number-of-free-parameters" title="Permalink to this headline">¶</a></h4>
<p>Reusing the optimised <code class="docutils literal notranslate"><span class="pre">lf</span></code> object from above, we can get the log-likelihood and the number of free parameters.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The number of free parameters (nfp) refers only to the number of parameters that were modifiable by the optimiser. Typically, the degrees-of-freedom of a likelihood ratio test statistic is computed as the difference in nfp between models. This will not be correct for models in which a boundary conditions exist (rate heterogeneity models where a parameter value boundary is set between bins).</p>
</div>
</div>
<div class="section" id="aikake-information-criterion">
<h4>Aikake Information Criterion<a class="headerlink" href="#aikake-information-criterion" title="Permalink to this headline">¶</a></h4>
<p>Reusing the optimised <code class="docutils literal notranslate"><span class="pre">lf</span></code> object from above.</p>
<p>We can also get the second-order AIC.</p>
</div>
<div class="section" id="bayesian-information-criterion">
<h4>Bayesian Information Criterion<a class="headerlink" href="#bayesian-information-criterion" title="Permalink to this headline">¶</a></h4>
<p>Reusing the optimised <code class="docutils literal notranslate"><span class="pre">lf</span></code> object from above.</p>
</div>
<div class="section" id="getting-maximum-likelihood-estimates">
<h4>Getting maximum likelihood estimates<a class="headerlink" href="#getting-maximum-likelihood-estimates" title="Permalink to this headline">¶</a></h4>
<p>Reusing the optimised <code class="docutils literal notranslate"><span class="pre">lf</span></code> object from above.</p>
<div class="section" id="one-at-a-time">
<h5>One at a time<a class="headerlink" href="#one-at-a-time" title="Permalink to this headline">¶</a></h5>
<p>We get the statistics out individually. We get the <code class="docutils literal notranslate"><span class="pre">length</span></code> for the Human edge and the exchangeability parameter <code class="docutils literal notranslate"><span class="pre">A/G</span></code>.</p>
</div>
<div class="section" id="just-the-motif-probabilities">
<h5>Just the motif probabilities<a class="headerlink" href="#just-the-motif-probabilities" title="Permalink to this headline">¶</a></h5>
</div>
<div class="section" id="as-tables">
<h5>As tables<a class="headerlink" href="#as-tables" title="Permalink to this headline">¶</a></h5>
</div>
</div>
</div>
<div class="section" id="testing-hypotheses-using-likelihood-ratio-tests">
<h3>Testing Hypotheses - Using Likelihood Ratio Tests<a class="headerlink" href="#testing-hypotheses-using-likelihood-ratio-tests" title="Permalink to this headline">¶</a></h3>
<p>We test the molecular clock hypothesis for human and chimpanzee lineages. The null has these two branches constrained to be equal.</p>
<p>The alternate allows the human and chimpanzee branches to differ by just setting all lengths to be independent.</p>
<p>We import the function for computing the probability of a chi-square test statistic, compute the likelihood ratio test statistic, degrees of freedom and the corresponding probability.</p>
<div class="section" id="testing-hypotheses-by-parametric-bootstrapping">
<h4>Testing Hypotheses - By parametric bootstrapping<a class="headerlink" href="#testing-hypotheses-by-parametric-bootstrapping" title="Permalink to this headline">¶</a></h4>
<p>If we can’t rely on the asymptotic behaviour of the LRT, e.g. due to small alignment length, we can use a parametric bootstrap. Convenience functions for that are described in more detail here <a class="reference internal" href="../examples/parametric_bootstrap.html#parametric-bootstrap"><span class="std std-ref">Performing a parametric bootstrap</span></a>.</p>
<p>In general, however, this capability derives from the ability of any defined <code class="docutils literal notranslate"><span class="pre">evolve</span></code> likelihood function to simulate an alignment. This property is provided as <code class="docutils literal notranslate"><span class="pre">simulate_alignment</span></code> method on likelihood function objects.</p>
</div>
</div>
<div class="section" id="determining-confidence-intervals-on-mles">
<h3>Determining confidence intervals on MLEs<a class="headerlink" href="#determining-confidence-intervals-on-mles" title="Permalink to this headline">¶</a></h3>
<p>The profile method is used to calculate a confidence interval for a named parameter. We show it here for a global substitution model exchangeability parameter (<em>kappa</em>, the ratio of transition to transversion rates) and for an edge specific parameter (just the human branch length).</p>
</div>
<div class="section" id="saving-results">
<h3>Saving results<a class="headerlink" href="#saving-results" title="Permalink to this headline">¶</a></h3>
<p>The best approach is to use the json string from the <code class="docutils literal notranslate"><span class="pre">to_json()</span></code> method. The saved data can be later reloaded using <code class="docutils literal notranslate"><span class="pre">cogent3.util.deserialise.deserialise_object()</span></code>. The <code class="docutils literal notranslate"><span class="pre">json</span></code> data contains the alignment, tree topology, substitution model, parameter values, etc..</p>
<p>To illustrate this, I create a very simple likelihood function. The <code class="docutils literal notranslate"><span class="pre">json</span></code> variable below is just a string that can be saved to disk.</p>
<p>We deserialise the object from the string.</p>
</div>
<div class="section" id="reconstructing-ancestral-sequences">
<h3>Reconstructing ancestral sequences<a class="headerlink" href="#reconstructing-ancestral-sequences" title="Permalink to this headline">¶</a></h3>
<p>We first fit a likelihood function.</p>
<p>We then get the most likely ancestral sequences.</p>
<p>Or we can get the posterior probabilities (returned as a <code class="docutils literal notranslate"><span class="pre">DictArray</span></code>) of sequence states at each node.</p>
<hr class="docutils" />
<p>There’s nothing that improves performance quite like being close to the maximum likelihood values. So using the <code class="docutils literal notranslate"><span class="pre">set_param_rule</span></code> method to provide good starting values can be very useful. As this can be difficult to do one easy way is to build simpler models that are nested within the one you’re interested in. Fitting those models and then relaxing constraints until you’re at the parameterisation of interest can markedly improve optimisation speed.</p>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020, cogent3 Team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.1.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>